{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stored 'COLORS' (dict)\n",
      "Stored 'SHORTEN' (dict)\n",
      "Stored 'TISSUES' (list)\n"
     ]
    }
   ],
   "source": [
    "# Colors (use colors from GTeX e.g. http://science.sciencemag.org/content/348/6235/648.full)\n",
    "COLORS = {\n",
    "    \"Artery-Aorta\":\"salmon\",\n",
    "    \"Artery-Tibial\": \"red\",\n",
    "    \"Adipose-Subcutaneous\": \"darkorange\",    \n",
    "    \"Adipose-Visceral\":\"orange\",\n",
    "    \"Brain-Caudate\":\"lemonchiffon\"   , \n",
    "    \"Brain-Cerebellum\":\"yellow\",\n",
    "    \"Cells-Transformedfibroblasts\": \"skyblue\",\n",
    "    \"Esophagus-Mucosa\": \"sienna\",\n",
    "    \"Esophagus-Muscularis\":\"burlywood\",\n",
    "    \"Heart-LeftVentricle\":\"darkviolet\",\n",
    "    \"Lung\": \"greenyellow\",\n",
    "    \"Muscle-Skeletal\": \"mediumslateblue\",\n",
    "    \"Nerve-Tibial\":\"gold\",\n",
    "    \"Skin-NotSunExposed\":\"blue\",\n",
    "    \"Skin-SunExposed\":\"cornflowerblue\",\n",
    "    \"Thyroid\":\"green\",\n",
    "    \"WholeBlood\": \"m\",\n",
    "    \"permuted\": \"gray\"\n",
    "}\n",
    "\n",
    "#    \"Thyroid\": \"green\",\n",
    "SHORTEN = {\n",
    "    \"Artery-Aorta\":\"Artery.A\"     ,\n",
    "    \"Artery-Tibial\": \"Artery.T\",\n",
    "    \"Adipose-Subcutaneous\": \"Adipose.S\",    \n",
    "    \"Adipose-Visceral\":\"Adipose.V\",\n",
    "    \"Brain-Caudate\":\"Caudate\"   , \n",
    "    \"Brain-Cerebellum\":\"Cerebellum\",\n",
    "    \"Cells-Transformedfibroblasts\": \"Fibroblast\",\n",
    "    \"Esophagus-Mucosa\": \"Mucosa\",\n",
    "    \"Esophagus-Muscularis\":\"Muscularis\",\n",
    "    \"Heart-LeftVentricle\":\"Heart\",\n",
    "    \"Lung\": \"Lung\",\n",
    "    \"Muscle-Skeletal\": \"Muscle\",\n",
    "    \"Nerve-Tibial\":\"Nerve\",\n",
    "    \"Skin-NotSunExposed\": \"SkinUnexposed\",\n",
    "    \"Skin-SunExposed\":\"SkinLeg\",\n",
    "    \"Thyroid\":\"Thyroid\",\n",
    "    \"WholeBlood\": \"Blood\",\n",
    "    \"permuted\":\"Permuted\",\n",
    "    \"LCL\": \"LCL\"\n",
    "}\n",
    "TISSUES = [item for item in list(COLORS.keys()) if item != \"permuted\"]\n",
    "\n",
    "def ExtractData(gene, chrom, start, tissue):\n",
    "    if \"(\" in tissue: tissue = tissue.split(\"(\")[0]\n",
    "    vcf = \"/storage/szfeupe/Runs/650GTEx_estr/Merged_STRs_all_samples.vcf.gz\"\n",
    "    # Pull out STR genotypes\n",
    "    cmd = \"\"\"bcftools query -r %s:%s-%s -f\"[%%SAMPLE\\\\t%%GB\\\\n]\" %s | \\\n",
    "    grep -v \"\\.\" | sed 's/|/\\\\t/' | awk '{print $1 \"\\\\t\" $2+$3}' > str_genotypes.tab\"\"\"%(chrom, start, start, vcf)\n",
    "    os.system(cmd)\n",
    "    # Pull out STR genotypes - alleles\n",
    "    cmd = \"\"\"bcftools query -r %s:%s-%s -f\"[%%SAMPLE\\\\t%%GB\\\\n]\" %s | \\\n",
    "    grep -v \"\\.\" | sed 's/|/\\\\t/' | awk '{print $1 \"\\\\t\" $2\",\"$3}' > str_genotypes_alleles.tab\"\"\"%(chrom, start, start, vcf)\n",
    "    os.system(cmd)\n",
    "    # Pull out gene expression\n",
    "    expr=\"/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/%s/Corr_Expr.csv\"%tissue\n",
    "    colnum = open(expr,\"r\").readline().split(\",\").index('\"' + gene + '\"')+2\n",
    "    cmd = \"\"\"cat %s | cut -d',' -f 1,%s | grep -v ENSG > expr.tab\"\"\"%(expr, colnum)\n",
    "    os.system(cmd)\n",
    "\n",
    "def PlotESTR(gene, tissue, chrom, start, COLORS):\n",
    "    strgt = pd.read_csv(\"str_genotypes.tab\", sep=\"\\t\", names=[\"sample\",\"str\"])\n",
    "    strgt[\"sample\"] = strgt[\"sample\"].apply(lambda x: \"-\".join(x.split(\"-\")[0:2]))\n",
    "    \n",
    "    rmgts = []\n",
    "    for gt in sorted(list(set(strgt[\"str\"]))):\n",
    "        if strgt[strgt[\"str\"]==gt].shape[0] <3: rmgts.append(gt)\n",
    "    strgt = strgt[strgt[\"str\"].apply(lambda x: x not in rmgts)]\n",
    "        \n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111)\n",
    "    expr = pd.read_csv(\"expr.tab\", names=[\"sample\",\"expr\"])\n",
    "    data = pd.merge(strgt, expr)\n",
    "    sns.swarmplot(x=\"str\", y=\"expr\", ax=ax, data=data, color=COLORS[tissue])\n",
    "    ax.set_xlabel(\"bp (rel. to hg19)\", size=15)\n",
    "    ax.set_ylabel(\"Expression - %s\"%tissue, size=15)\n",
    "    ax.set_xticklabels([int(item) for item in sorted(list(set(data[\"str\"])))], size=12)\n",
    "    ax.set_yticklabels([\"%.2f\"%(item) for item in ax.get_yticks()], size=12)\n",
    "    ax.set_title(\"\")\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.xaxis.set_ticks_position('bottom')\n",
    "    ax.set_title(\"chr%s:%s - %s\"%(chrom, start, gene))\n",
    "    plt.suptitle(\"\");\n",
    "    plt.show()\n",
    "    fig.savefig(\"%s_estr.pdf\"%gene)\n",
    "\n",
    "def PlotHeatmap():\n",
    "    strgt = pd.read_csv(\"str_genotypes_alleles.tab\", sep=\"\\t\", names=[\"sample\",\"GB\"])\n",
    "    strgt[\"sample\"] = strgt[\"sample\"].apply(lambda x: \"-\".join(x.split(\"-\")[0:2]))\n",
    "    strgt[\"str\"] = strgt[\"GB\"].apply(lambda x: sum([int(item) for item in x.split(\",\")]))\n",
    "    rmgts = []\n",
    "    for gt in sorted(list(set(strgt[\"str\"]))):\n",
    "        if strgt[strgt[\"str\"]==gt].shape[0] <3: rmgts.append(gt)\n",
    "    strgt = strgt[strgt[\"str\"].apply(lambda x: x not in rmgts)]\n",
    "\n",
    "    expr = pd.read_csv(\"expr.tab\", names=[\"sample\",\"expr\"])\n",
    "    correxpr = pd.merge(strgt, expr)\n",
    "    # Heatmap\n",
    "    alleles = set()\n",
    "    for i in range(correxpr.shape[0]):\n",
    "        a1, a2 = correxpr[\"GB\"].values[i].split(\",\")\n",
    "        alleles.add(int(a1))\n",
    "        alleles.add(int(a2))\n",
    "    alleles = sorted(list(alleles))\n",
    "\n",
    "    exprmatrix = np.zeros((len(alleles), len(alleles)))\n",
    "    counts = np.zeros((len(alleles), len(alleles)))\n",
    "    for i in range(correxpr.shape[0]):\n",
    "        a1, a2 = correxpr[\"GB\"].values[i].split(\",\")\n",
    "        a1 = int(a1)\n",
    "        a2 = int(a2)\n",
    "        if a1<a2:\n",
    "            a1ind = alleles.index(a1)\n",
    "            a2ind = alleles.index(a2)\n",
    "        else: \n",
    "            a1ind = alleles.index(a2)\n",
    "            a2ind = alleles.index(a1)\n",
    "        exprmatrix[a1ind,a2ind] += correxpr[\"expr\"].values[i]\n",
    "        exprmatrix[a2ind,a1ind] += correxpr[\"expr\"].values[i]\n",
    "        counts[a1ind,a2ind] += 1\n",
    "        counts[a2ind,a1ind] += 1\n",
    "    exprmatrix = exprmatrix/(counts+1)\n",
    "\n",
    "    sns.heatmap(exprmatrix, xticklabels=alleles, yticklabels=alleles)\n",
    "    \n",
    "    # Plot each row of the matrix\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111)\n",
    "    for i in range(exprmatrix.shape[0]):\n",
    "        vals = exprmatrix[i,:]\n",
    "        if sum(counts[i,:]) < 25: continue\n",
    "        nvals = []\n",
    "        for item in vals:\n",
    "            if item == 0: nvals.append(None)\n",
    "            else: nvals.append(item)\n",
    "        ax.plot(alleles, nvals, label=alleles[i])\n",
    "    ax.legend()\n",
    "\n",
    "%store COLORS\n",
    "%store SHORTEN\n",
    "%store TISSUES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
