For running notebooks you should run the the notebook named gtex_variables to initalize the colors and tissues short names first.
OR
initalize them directly.

Figure1         qqplot
                Effect size comparison
                pairwise tissue comparison
                unique vs shared ***

Figure 2 & 2': Causality Analysis
        1.	Distribution of CAVIAR scores (let’s see how it looks) 
        2.	Example “causal loci”. Color dots by genotype of best SNP 
        1.	Multi-allelic example from ASHG poster 
        2.	CSTB example 
        3.	One “negative” example of eSTR with good p-value but poor CAVIAR score 
        3.	Enrichment as a function of distance to TSS, TES, nearest splice site  
        4.	Enrichment by motif 

Figure 3 & X :Examples with biological insights 
        (1) eSTRs effect on RNA Binding Proteins (RBPs) 
        (2) eSTRs that overlap RBPs 
        (3)eSTRs effects on transcription factors (TFs) 
        (4)eSTRs that overlap transcrion factors
        
Figure4: Analyzing non linear regression
    (1) Number of eSTRs using non linear regression
    (2) Particular observation 


Figure5: Pull out some GWAS examples 
        (1) plot expression vs the STR genotype
        (2) plot LOcus zoom


                 Supp Figures
                 
Supplementary FigureX: This has to do with homopolymers. Need to look at the homopolymers specifically
                including their locatiopn, their occurence and length, possible biological insights

Supplementary Figure 2:       SNP PCA with 1000 Genomes samples

Supplementary Figure 1:       Profiling STRs in GTEx samples Number of calls/sample Number of samples/locus Heterozygosity vs. previously reported in SGDP () Stutter model summary

Supplementary Figure 3:       PEER factor analysis heatmap

Supplementary Figure 4:       Unique vs. shared estrs per tissue

Supplementary Table 1 :       Litterature Examples


#RNA binding proteins    #transcription factors

