{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "A. For each tissue, enrichment of each motif in \"eSTRs\" vs. all analyzed STRs as background. Just a simple fisher exact test for this.\n",
    "\"\"\"\n",
    "\n",
    "# Libraries\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scipy.stats as ss\n",
    "import seaborn.apionly as sns\n",
    "\n",
    "COLORS = {\n",
    "    \"Artery-Aorta\":\"salmon\",\n",
    "    \"Artery-Tibial\": \"red\",\n",
    "    \"Adipose-Subcutaneous\": \"darkorange\",    \n",
    "    \"Adipose-Visceral(Omentum)\":\"orange\",\n",
    "    \"Brain-Caudate(basalganglia)\":\"lemonchiffon\"   , \n",
    "    \"Brain-Cerebellum\":\"yellow\",\n",
    "    \"Cells-Transformedfibroblasts\": \"skyblue\",\n",
    "    \"Esophagus-Mucosa\": \"sienna\",\n",
    "    \"Esophagus-Muscularis\":\"burlywood\",\n",
    "    \"Heart-LeftVentricle\":\"darkviolet\",\n",
    "    \"Lung\": \"greenyellow\",\n",
    "    \"Muscle-Skeletal\": \"mediumslateblue\",\n",
    "    \"Nerve-Tibial\":\"gold\",\n",
    "    \"Skin-NotSunExposed(Suprapubic)\":\"blue\",\n",
    "    \"Skin-SunExposed(Lowerleg)\":\"cornflowerblue\",\n",
    "    \"Thyroid\":\"green\",\n",
    "    \"WholeBlood\": \"m\",\n",
    "    \"permuted\": \"gray\"\n",
    "}\n",
    "\n",
    "#    \"Thyroid\": \"green\",\n",
    "SHORTEN = {\n",
    "    \"Artery-Aorta\":\"Artery.A\"     ,\n",
    "    \"Artery-Tibial\": \"Artery.T\",\n",
    "    \"Adipose-Subcutaneous\": \"Adipose.S\",    \n",
    "    \"Adipose-Visceral(Omentum)\":\"Adipose.V\",\n",
    "    \"Brain-Caudate(basalganglia)\":\"Caudate\"   , \n",
    "    \"Brain-Cerebellum\":\"Cerebellum\",\n",
    "    \"Cells-Transformedfibroblasts\": \"Fibroblast\",\n",
    "    \"Esophagus-Mucosa\": \"Mucosa\",\n",
    "    \"Esophagus-Muscularis\":\"Muscularis\",\n",
    "    \"Heart-LeftVentricle\":\"Ventricule\",\n",
    "    \"Lung\": \"Lung\",\n",
    "    \"Muscle-Skeletal\": \"Muscle\",\n",
    "    \"Nerve-Tibial\":\"Nerve\",\n",
    "    \"Skin-NotSunExposed(Suprapubic)\": \"SkinUnexposed\",\n",
    "    \"Skin-SunExposed(Lowerleg)\":\"SkinLeg\",\n",
    "    \"Thyroid\":\"Thyroid\",\n",
    "    \"WholeBlood\": \"Blood\",\n",
    "    \"permuted\":\"Permuted\",\n",
    "    \"LCL\": \"LCL\"\n",
    "}\n",
    "\n",
    "def Findoverlap(Frame, feat ):\n",
    "    L=sorted(set(list(Frame['chrom'])))\n",
    "    fragments=[]\n",
    "    t=0\n",
    "    genes=[]\n",
    "    for C in L:\n",
    "        X = Frame.loc[Frame['chrom']==C]\n",
    "        Y = feat.loc[feat['chrom']==C]\n",
    "        #print(X.shape, Y.shape)\n",
    "        X['str.start'] = X[\"str.start\"].astype(int)\n",
    "        L1 = list(X['str.start'])\n",
    "        L2=[]\n",
    "        for start in L1:\n",
    "            X2 = Y.loc[(Y[\"start\"]<=start) & (Y[\"stop\"]>=start)]\n",
    "            if X2.shape[0]==0:\n",
    "                continue\n",
    "            else:\n",
    "                L2.append(start)        \n",
    "        fragments.append(X.loc[X['str.start'].isin(L2)])\n",
    "        #print(C,'\\t',fragments[-1].shape)\n",
    "    result = pd.concat(fragments)\n",
    "    print(Frame.shape, '\\t', result.shape)\n",
    "    return(result)\n",
    "\n",
    "TISSUES = [item for item in list(COLORS.keys()) if item != \"permuted\"]\n",
    "\n",
    "regr='/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/LR_SummaryTest_Table.tsv'\n",
    "feat='/storage/szfeupe/Runs/GTEx_estr/FEATURES/Allgencodefeatures_table'\n",
    "motif='/storage/resources/dbase/human/hg19/hg19.hipstr_reference_withmotif.bed'\n",
    "Reg1 = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/Merged_Best_causality.Table'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "F=[] ; G=[] ; R = []\n",
    "#open regression\n",
    "R = pd.read_csv(regr, sep='\\t')  ; X = R[[\"gene\",\"chrom\",\"str.start\",\"E.tissues\",\"motif\"]]\n",
    "X['best.str.start'] = X['str.start']\n",
    "#open Causal annot\n",
    "C = pd.read_csv(Reg1, sep='\\t')\n",
    "G=pd.merge(C,X, on=['gene','chrom','best.str.start'])\n",
    "#open annotation\n",
    "FT = pd.read_csv(feat, sep='\\t')\n",
    "FTGene = FT.loc[FT['feature.type']=='gene']\n",
    "G =Findoverlap(G,FTGene)\n",
    "\n",
    "##@ Causal STR\n",
    "data = G\n",
    "#data1 = G[G['top.variant'].str.contains(\"STR_\")==True]\n",
    "#data = G.loc[G['E.tissues']>0]\n",
    "#data_e = data1.loc[data1[\"E.tissues\"]>0]\n",
    "data1 = data[data['top.variant'].str.contains(\"STR_\")==True]\n",
    "data_e = data1.loc[data1[\"E.tissues\"]>0]\n",
    "#Enrichment\n",
    "\n",
    "Estr = data_e.shape[0]\n",
    "Str = data.shape[0]\n",
    "Motifs = list(set(list(data_e['motif'])))\n",
    "Mo_counts=[list(data_e['motif']).count(x) for x in Motifs]\n",
    "All = []\n",
    "sig_motifs=[]\n",
    "for M in Motifs:\n",
    "    data_eM = data_e.loc[data_e['motif']==M]\n",
    "    data_M = data.loc[data['motif']==M]\n",
    "    Mestrs = data_eM.shape[0]\n",
    "    Mstrs = data_M.shape[0]\n",
    "    CTT = [[Mestrs, Estr - Mestrs] , [Mstrs-Mestrs , Str-Estr -(Mstrs - Mestrs)]] #contengency\n",
    "    oddsratio, pv = ss.fisher_exact([CTT[0], CTT[1]])\n",
    "    All.append([M,list(data_e['motif']).count(M),-np.log10(pv),oddsratio])\n",
    "    if pv<=0.05:\n",
    "        sig_motifs.append([M,pv*100,oddsratio])\n",
    "p=Estr/data.shape[0]\n",
    "\n",
    "#print(\"Likelihood ratio of an'AC-rich' eSTR is \",(data_eac.shape[0]/data_e.shape[0])/((data_ac.shape[0]-data_eac.shape[0])/(data_ac.shape[0])) )\n",
    "\n",
    "count_uniq_mo = sorted(list(set([x[2] for x in All])))\n",
    "ordered_All = []\n",
    "for u in count_uniq_mo:\n",
    "    ordered_All = ordered_All + [x for x in All if x[2]==u] \n",
    "\n",
    "\n",
    "plt.figure(figsize=(15,5))\n",
    "Y=[x[2] for x in ordered_All]\n",
    "X=[x for x in range(len(ordered_All))]\n",
    "#Y = Y[:-1] ; X = X[:-1]\n",
    "col = 'b' #[COLORS[tis] if x>-np.log10(0.05) else 'k' for x in Y ]\n",
    "#plt.scatter(X, Y, c=col, s=20, linewidth=1)\n",
    "plt.bar(X, Y, color=col)\n",
    "plt.xlabel('Motifs')\n",
    "plt.ylabel('-log10(pvalues) motif enrichment')\n",
    "plt.title('Causal eSTRs Motif enrichment (#eSTRs='+str(data_e.shape[0])+') vs all gene STRs')\n",
    "plt.xticks(X,[x[0] for x in ordered_All], rotation='vertical')\n",
    "plt.axhline(y=-np.log10(0.05), xmin=0, xmax=1, hold=None, alpha=0.5, ls='--')\n",
    "#plt.axhline(y=-np.log10(0.1), xmin=0, xmax=1, hold=None, alpha=0.5, ls='--')\n",
    "plt.text(1, 1.5, 'pvalue=0.05', color='grey')\n",
    "plt.show()  \n",
    "    \n",
    "#plt.savefig(\"motifsE.png\")   \n",
    "\n",
    "\"\"\"\n",
    "Signigicantly enriched motifs\n",
    "'AGATAT'*, 'AAAG', 'AAATG', 'AGGGGG', 'ACGCC', 'AGAG', 'AACC', 'AAAAG', 'AATG', 'AAAT', 'AT', 'AC'\n",
    "\"\"\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
