{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import math\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import random\n",
    "import shutil\n",
    "import sys\n",
    "import gzip\n",
    "\n",
    "def PROGRESS(msg, printit=True):\n",
    "    if printit: # false for some messages when not in debug mode\n",
    "        sys.stderr.write(\"%s\\n\"%msg.strip())\n",
    "        \n",
    "def MakeZScoreTable(vals):\n",
    "    try:\n",
    "        Zscore=vals['beta']/vals['beta.se']\n",
    "    except:\n",
    "        Zscore=None\n",
    "    return Zscore\n",
    "\n",
    "def WriteCorrTable(indexed_genotypes):\n",
    "    \"\"\" generate correlation table using normalized genotype\n",
    "      _1_ ... _n_\n",
    "    1| 1  ... C1n\n",
    "    .|    ...\n",
    "    n|Cn1 ... Cnn=1\n",
    "    \"\"\"\n",
    "    G=indexed_genotypes.transpose()\n",
    "    variants = list(G.columns)\n",
    "    CMat=[]\n",
    "    print ('\\t\\t\\t**', len(variants))\n",
    "    for V1 in variants:\n",
    "        COV=[]\n",
    "        for V2 in variants:\n",
    "            X=G[V1].replace('None', np.nan).astype(float)\n",
    "            Y=G[V2].replace('None', np.nan).astype(float)\n",
    "            #COV.append(X.corr(Y))\n",
    "            if X.corr(Y) is np.nan:    #### \n",
    "                COV.append(0.0) #For missing LD we assume non linear corr (undetermined LD)\n",
    "            else:\n",
    "                COV.append(X.corr(Y))\n",
    "        CMat.append(COV)\n",
    "    return pd.DataFrame(CMat,columns=variants, index=variants) \n",
    "\n",
    "PATH = '/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/Adipose-Subcutaneous/'\n",
    "EXPRFILE = PATH+'Corr_Expr.csv'\n",
    "EXPRANNOTFILE = '/storage/resources/dbase/human/hg19/gencode_gene_annotations_hg19.csv'\n",
    "CHROM = 22\n",
    "if \"chr\" not in str(CHROM): CHROM=\"chr%s\"%CHROM\n",
    "DISTFROMGENE = 100000\n",
    "STRGTFILE = '/storage/szfeupe/Runs/650GTEx_estr/Genotypes/GenotypesNormalized.table'\n",
    "SNPGTFILE = '/storage/szfeupe/Runs/650GTEx_estr/SNP_Analysis/chr22.tab'\n",
    "OUTFILE = 'help.tab'\n",
    "REG_STRs = PATH+'/Lin_Reg_Out'\n",
    "REG_SNPs = PATH+'SNP_Analysis/Lin_Reg_Out'\n",
    "TMPDIR = PATH+'HH/caviar_temps/'\n",
    "if not os.path.exists(TMPDIR):\n",
    "    os.mkdir(TMPDIR)\n",
    "DEBUG =True\n",
    "ESTRGENESFILE=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "    # Load expression\n",
    "    DEBUG =True\n",
    "    PROGRESS(\"\\nLoad expression\", printit=DEBUG)\n",
    "    expr = pd.read_csv(EXPRFILE)\n",
    "    samples_to_keep = list(expr.index)\n",
    "    # Load annotation\n",
    "    PROGRESS(\"Load annotation\", printit=DEBUG)\n",
    "    expr_annot = pd.read_csv(EXPRANNOTFILE)\n",
    "    expr_annot.index = expr_annot[\"probe.id\"].values\n",
    "    expr_annot = expr_annot.reindex(list(expr.columns))\n",
    "    expr_annot = expr_annot.dropna() \n",
    "    expr_annot = expr_annot[expr_annot[\"gene.chr\"] == CHROM]\n",
    "    # Load strs Regression\n",
    "    PROGRESS(\"\\nLoad strs regression\", printit=DEBUG)\n",
    "    strs = pd.read_csv(REG_STRs, sep=\"\\t\")\n",
    "    strs = strs.loc[strs['chrom']==CHROM]\n",
    "    # Load snps regression\n",
    "    PROGRESS(\"\\nLoad snps regression\", printit=DEBUG)\n",
    "    snps = pd.read_csv(REG_SNPs, sep=\"\\t\")\n",
    "    snps = snps.loc[snps['chrom']==CHROM]\n",
    "    del snps['Unnamed: 0']\n",
    "    #Load SNP genotypes\n",
    "    PROGRESS(\"Load SNPs\", printit=DEBUG)\n",
    "    snpgt = pd.read_csv(SNPGTFILE, sep=\"\\t\",low_memory=False)\n",
    "    snpgt = snpgt.loc[snpgt['chrom']==CHROM]\n",
    "    # Load STR genotypes\n",
    "    PROGRESS(\"Load STRs\", printit=DEBUG)\n",
    "    strgt = pd.read_csv(STRGTFILE, sep=\"\\t\")\n",
    "    strgt = strgt.loc[strgt['chrom']==CHROM]\n",
    "    # Restrict to STR samples\n",
    "    PROGRESS(\"Restrict to STRs samples\", printit=DEBUG)\n",
    "    str_samples = samples_to_keep\n",
    "    expr = expr.reindex(str_samples)\n",
    "    snpgt = snpgt[[\"chrom\",\"start\"] + str_samples]\n",
    "    snpgt.index = list(snpgt[\"start\"].apply(lambda x: \"SNP_%s\"%int(x)))\n",
    "    strgt = strgt[[\"chrom\",\"start\"] + str_samples]\n",
    "    strgt.index = list(strgt[\"start\"].apply(lambda x: \"STR_%s\"%int(x)))\n",
    "    # Load eSTR results\n",
    "    PROGRESS(\"Restrict to eSTR genes only\", printit=DEBUG)\n",
    "    if ESTRGENESFILE is not None:\n",
    "        estr_genes = pd.read_csv(ESTRGENESFILE, sep=\"\\t\")\n",
    "        Genes = estr_genes.loc[estr_genes['qvalue']<=0.1]['gene']  # estrs at 10%FDR\n",
    "        expr_annot = expr_annot.loc[expr_annot['gene.id'].isin(list(Genes))]\n",
    "    print (expr_annot.shape)\n",
    "    #open output files\n",
    "    Errorfile = open(TMPDIR+\"/Errorfile.out\", 'w')\n",
    "    OUT = open(OUTFILE, \"w\")\n",
    "    OUT.write(\"\\t\".join([\"chrom\", \"gene\", \"best.str.start\", \"best.str.score\", \"top.variant\", \"top.variant.score\",\"top.snp.score\"])+'\\n')\n",
    "    # For each gene, get all cis-variants and the best STR\n",
    "    for i in range(expr_annot.shape[0]):\n",
    "        gene=expr_annot.index.values[i]\n",
    "        ensgene = expr_annot[\"gene.id\"].values[i]  #'ENSG00000215912.7'\n",
    "        genedir=TMPDIR+\"/%s\"%gene\n",
    "        if not os.path.exists(genedir):\n",
    "            os.mkdir(genedir)\n",
    "        PROGRESS(\"Getting data for %s\"%gene, printit=DEBUG)\n",
    "        start = expr_annot[\"gene.start\"].values[i]\n",
    "        end = expr_annot[\"gene.stop\"].values[i]\n",
    "    # Pull out cis SNPs\n",
    "        PROGRESS(\"Getting cis SNPs for %s\"%gene)\n",
    "        cis_snps = snps[(snps[\"str.start\"] >= (start-DISTFROMGENE)) & (snps[\"str.start\"] <= (end+DISTFROMGENE))]\n",
    "        #print cis_snps.shape , '###'\n",
    "        cis_snps = cis_snps.loc[cis_snps[\"str.start\"].isin(list(snpgt[\"start\"]))]  ###\n",
    "        cis_variants = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_variants=cis_variants.sort_values(by=\"p.wald\").head(n=100)\n",
    "        cis_variants.index = cis_variants[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        L=list(cis_variants.index)\n",
    "    # Pull out cis STR\n",
    "        PROGRESS(\"Getting most significant cis STR for %s\"%gene)\n",
    "        cis_strs = strs[strs[\"gene\"]==ensgene].sort_values(\"p.wald\")\n",
    "        if cis_strs.shape[0]==0:\n",
    "            PROGRESS(\"There are no STRs found for %s... Gene not in LR table\"%gene)\n",
    "            continue\n",
    "        else: \n",
    "            cis_strs.index = cis_strs[\"str.start\"].apply(lambda x: \"STR_%s\"%int(x)) \n",
    "            L0 = list(cis_strs.index)\n",
    "        #\n",
    "        cis_variants = pd.concat([cis_snps, cis_strs])\n",
    "    # Make z file data\n",
    "        Ztable = MakeZScoreTable(cis_variants[['beta','beta.se']])\n",
    "        if Ztable is None:\n",
    "            Errorfile.write(gene+\": Z score could not be calculated; beta.se is probably 0 or null\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            Ztable.to_csv(genedir+'/ZFILE', sep='\\t',header=None)\n",
    "    # Make LD file\n",
    "        genotypes = snpgt.loc[L]\n",
    "        genotypes = pd.concat([genotypes, strgt.loc[L0] ])\n",
    "        del genotypes['chrom']\n",
    "        del genotypes['start']\n",
    "        CorrMatrix = WriteCorrTable(genotypes)\n",
    "        CorrMatrix.to_csv(genedir+'/LDFILE', sep='\\t',header=None, index=None)\n",
    "        PROGRESS(\"Matrix of corr was sent to file for %s\"%gene)        \n",
    "    #Run caviar\n",
    "        caviar_cmd = \"CAVIAR -l %s -z %s -o %s/caviar -c 1 -f 1 > %s\"%(genedir+\"/LDFILE\", genedir+\"/ZFILE\", genedir, genedir+\"/log\")\n",
    "        os.system(caviar_cmd)\n",
    "    #Output results\n",
    "        if not os.path.exists(genedir+'/caviar_post'):\n",
    "            Errorfile.write(gene+\": CAVIAR did not run.\\n\\tERROR: Segmentation fault (core dumped) in log file\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            post = pd.read_csv(genedir+'/caviar_post', sep=\"\\t\", header=None)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Getting data for ENSG00000177663.9\n",
      "Getting cis SNPs for ENSG00000177663.9\n",
      "Getting most significant cis STR for ENSG00000177663.9\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t\t\t** 112\n",
      "ENSG00000177663.9 \tExpression annotation\t 17565844.0 17596583.0\n",
      "top 100 cis_snps\t ['SNP_17586609', 'SNP_17586715', '...', 'SNP_17569017'] \t  100 (100, 12)\n",
      "All cis strs\t (12, 12) #str#\n",
      "All cis variants\t (112, 12)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matrix of corr was sent to file for ENSG00000177663.9\n"
     ]
    }
   ],
   "source": [
    "    # For each gene, get all cis-variants and the best STR\n",
    "    for i in range(expr_annot.shape[0]):\n",
    "        gene=expr_annot.index.values[i]\n",
    "        ensgene = expr_annot[\"gene.id\"].values[i]  #'ENSG00000215912.7'\n",
    "        genedir=TMPDIR+\"/%s\"%gene\n",
    "        if not os.path.exists(genedir):\n",
    "            os.mkdir(genedir)\n",
    "        PROGRESS(\"Getting data for %s\"%gene, printit=DEBUG)\n",
    "        start = expr_annot[\"gene.start\"].values[i]\n",
    "        end = expr_annot[\"gene.stop\"].values[i]\n",
    "    # Pull out cis SNPs\n",
    "        PROGRESS(\"Getting cis SNPs for %s\"%gene)\n",
    "        cis_snps = snps[(snps[\"str.start\"] >= (start-DISTFROMGENE)) & (snps[\"str.start\"] <= (end+DISTFROMGENE))]\n",
    "        #print cis_snps.shape , '###'\n",
    "        cis_snps = cis_snps.loc[cis_snps[\"str.start\"].isin(list(snpgt[\"start\"]))]  ###\n",
    "        cis_variants = cis_snps.loc[cis_snps['gene']==ensgene]\n",
    "        cis_snps=cis_variants.sort_values(by=\"p.wald\").head(n=100)\n",
    "        cis_snps.index = cis_snps[\"str.start\"].apply(lambda x: \"SNP_%s\"%int(x))\n",
    "        L=list(cis_snps.index)\n",
    "    # Pull out cis STR\n",
    "        PROGRESS(\"Getting most significant cis STR for %s\"%gene)\n",
    "        cis_strs = strs[strs[\"gene\"]==ensgene].sort_values(\"p.wald\")\n",
    "        if cis_strs.shape[0]==0:\n",
    "            PROGRESS(\"There are no STRs found for %s... Gene not in LR table\"%gene)\n",
    "            continue\n",
    "        else: \n",
    "            cis_strs.index = cis_strs[\"str.start\"].apply(lambda x: \"STR_%s\"%int(x))\n",
    "            best_str_start = int(cis_strs[\"str.start\"].values[0])\n",
    "            L0 = list(cis_strs.index)\n",
    "        #\n",
    "        cis_variants = pd.concat([cis_snps, cis_strs])\n",
    "    # Make z file data\n",
    "        Ztable = MakeZScoreTable(cis_variants[['beta','beta.se']])\n",
    "        if Ztable is None:\n",
    "            Errorfile.write(gene+\": Z score could not be calculated; beta.se is probably 0 or null\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            Ztable.to_csv(genedir+'/ZFILE', sep='\\t',header=None)\n",
    "    # Make LD file\n",
    "        genotypes = snpgt.loc[L]\n",
    "        genotypes = pd.concat([genotypes, strgt.loc[L0] ])\n",
    "        del genotypes['chrom']\n",
    "        del genotypes['start']\n",
    "        CorrMatrix = WriteCorrTable(genotypes)\n",
    "        CorrMatrix.to_csv(genedir+'/LDFILE', sep='\\t',header=None, index=None)\n",
    "        PROGRESS(\"Matrix of corr was sent to file for %s\"%gene)        \n",
    "    #Run caviar\n",
    "        caviar_cmd = \"CAVIAR -l %s -z %s -o %s/caviar -c 1 -f 1 > %s\"%(genedir+\"/LDFILE\", genedir+\"/ZFILE\", genedir, genedir+\"/log\")\n",
    "        os.system(caviar_cmd)\n",
    "    #Output results\n",
    "        if not os.path.exists(genedir+'/caviar_post'):\n",
    "            Errorfile.write(gene+\": CAVIAR did not run.\\n\\tERROR: Segmentation fault (core dumped) in log file\\n\")\n",
    "            continue\n",
    "        else:\n",
    "            post = pd.read_csv(genedir+'/caviar_post', sep=\"\\t\", header=None)\n",
    "            #caviarstr  =  post.loc[post[0]=='STR_'+str(best_str_start)][2].tolist()[0]\n",
    "            #topvariant =  post.sort_values(post.columns[2], ascending=False).values[0][0]\n",
    "            #topscore  =  post.sort_values(post.columns[2], ascending=False).values[0][2]\n",
    "            \n",
    "        \n",
    "        print(gene, '\\tExpression annotation\\t', start, end)\n",
    "        print('top 100 cis_snps\\t', [L[0], L[1],'...',L[-1]],'\\t ',len(L), cis_snps.shape)\n",
    "        print('All cis strs\\t', cis_strs.shape , '#str#') \n",
    "        print('All cis variants\\t', cis_variants.shape)\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Load expression\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['GTEX-1117F', 'GTEX-111FC', 'GTEX-111VG', 'GTEX-111YS', 'GTEX-1122O',\n",
       "       'GTEX-117YX', 'GTEX-11DXZ', 'GTEX-11DZ1', 'GTEX-11EM3', 'GTEX-11EQ8',\n",
       "       ...\n",
       "       'GTEX-ZVZQ', 'GTEX-ZXES', 'GTEX-ZXG5', 'GTEX-ZYFC', 'GTEX-ZYFD',\n",
       "       'GTEX-ZYVF', 'GTEX-ZYW4', 'GTEX-ZYWO', 'GTEX-ZZ64', 'GTEX-ZZPU'],\n",
       "      dtype='object', length=270)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "        else:\n",
    "            post = pd.read_csv(genedir+'/caviar_post', sep=\"\\t\", header=None)\n",
    "            print 'post','......', post    #.loc[post[0]=='STR_'+str(best_str_start)][2].tolist()\n",
    "            #caviarstr  =  post.loc[post[0]=='STR_'+str(best_str_start)][2].tolist()[0]\n",
    "            #topvariant =  post.sort_values(post.columns[2], ascending=False).values[0][0]\n",
    "            #topscore  =  post.sort_values(post.columns[2], ascending=False).values[0][2]\n",
    "            #if 'STR' in topvariant:\n",
    "            #    snpscore  =  post.sort_values(post.columns[2], ascending=False).values[1][2]\n",
    "            #else:\n",
    "            #    snpscore=topscore\n",
    "            #OUT.write(\"\\t\".join([CHROM, gene, str(best_str_start), str(caviarstr), topvariant, str(topscore), str(snpscore)])+'\\n')\n",
    "        O=0\n",
    "        break\n",
    "    OUT.close()\n",
    "    Errorfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNP_17586609</td>\n",
       "      <td>7.649090e-01</td>\n",
       "      <td>7.649090e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNP_17586715</td>\n",
       "      <td>1.064020e-01</td>\n",
       "      <td>1.064020e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNP_17586631</td>\n",
       "      <td>1.048040e-01</td>\n",
       "      <td>1.048040e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNP_17586583</td>\n",
       "      <td>2.388540e-02</td>\n",
       "      <td>2.388540e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNP_17602164</td>\n",
       "      <td>7.068190e-11</td>\n",
       "      <td>7.068190e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SNP_17595915</td>\n",
       "      <td>4.266680e-11</td>\n",
       "      <td>4.266680e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SNP_17595929</td>\n",
       "      <td>4.266680e-11</td>\n",
       "      <td>4.266680e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SNP_17591089</td>\n",
       "      <td>2.184990e-11</td>\n",
       "      <td>2.184990e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SNP_17596322</td>\n",
       "      <td>1.692930e-11</td>\n",
       "      <td>1.692930e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SNP_17601466</td>\n",
       "      <td>5.988680e-12</td>\n",
       "      <td>5.988680e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SNP_17582788</td>\n",
       "      <td>3.378290e-13</td>\n",
       "      <td>3.378290e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SNP_17591144</td>\n",
       "      <td>2.538860e-13</td>\n",
       "      <td>2.538860e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SNP_17585808</td>\n",
       "      <td>7.207200e-14</td>\n",
       "      <td>7.207200e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SNP_17591697</td>\n",
       "      <td>4.957880e-14</td>\n",
       "      <td>4.957880e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SNP_17605197</td>\n",
       "      <td>4.727070e-14</td>\n",
       "      <td>4.727070e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SNP_17595070</td>\n",
       "      <td>2.404580e-14</td>\n",
       "      <td>2.404580e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SNP_17594886</td>\n",
       "      <td>1.983560e-14</td>\n",
       "      <td>1.983560e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SNP_17594915</td>\n",
       "      <td>1.983560e-14</td>\n",
       "      <td>1.983560e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SNP_17582044</td>\n",
       "      <td>5.325070e-15</td>\n",
       "      <td>5.325070e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SNP_17586471</td>\n",
       "      <td>2.582190e-15</td>\n",
       "      <td>2.582190e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SNP_17587515</td>\n",
       "      <td>2.334290e-15</td>\n",
       "      <td>2.334290e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SNP_17582061</td>\n",
       "      <td>2.159920e-15</td>\n",
       "      <td>2.159920e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SNP_17582776</td>\n",
       "      <td>9.317800e-16</td>\n",
       "      <td>9.317800e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SNP_17587746</td>\n",
       "      <td>9.254720e-16</td>\n",
       "      <td>9.254720e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SNP_17583492</td>\n",
       "      <td>8.901370e-16</td>\n",
       "      <td>8.901370e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SNP_17585465</td>\n",
       "      <td>8.301790e-16</td>\n",
       "      <td>8.301790e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SNP_17599035</td>\n",
       "      <td>5.337240e-16</td>\n",
       "      <td>5.337240e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SNP_17589152</td>\n",
       "      <td>5.171440e-16</td>\n",
       "      <td>5.171440e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SNP_17606639</td>\n",
       "      <td>4.790680e-16</td>\n",
       "      <td>4.790680e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SNP_17623675</td>\n",
       "      <td>4.737600e-16</td>\n",
       "      <td>4.737600e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>SNP_17566504</td>\n",
       "      <td>4.759770e-17</td>\n",
       "      <td>4.759770e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SNP_17603801</td>\n",
       "      <td>4.700440e-17</td>\n",
       "      <td>4.700440e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>SNP_17584677</td>\n",
       "      <td>4.499790e-17</td>\n",
       "      <td>4.499790e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>SNP_17525584</td>\n",
       "      <td>4.443330e-17</td>\n",
       "      <td>4.443330e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>SNP_17598512</td>\n",
       "      <td>4.311950e-17</td>\n",
       "      <td>4.311950e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SNP_17525970</td>\n",
       "      <td>4.120170e-17</td>\n",
       "      <td>4.120170e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>SNP_17589209</td>\n",
       "      <td>3.956490e-17</td>\n",
       "      <td>3.956490e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>SNP_17597056</td>\n",
       "      <td>3.644230e-17</td>\n",
       "      <td>3.644230e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>SNP_17609330</td>\n",
       "      <td>3.611320e-17</td>\n",
       "      <td>3.611320e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>SNP_17641500</td>\n",
       "      <td>3.582020e-17</td>\n",
       "      <td>3.582020e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>SNP_17668850</td>\n",
       "      <td>3.527790e-17</td>\n",
       "      <td>3.527790e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>SNP_17668851</td>\n",
       "      <td>3.527790e-17</td>\n",
       "      <td>3.527790e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>SNP_17567699</td>\n",
       "      <td>3.245120e-17</td>\n",
       "      <td>3.245120e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>SNP_17651273</td>\n",
       "      <td>3.121340e-17</td>\n",
       "      <td>3.121340e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>SNP_17650931</td>\n",
       "      <td>3.121340e-17</td>\n",
       "      <td>3.121340e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>SNP_17619378</td>\n",
       "      <td>3.082660e-17</td>\n",
       "      <td>3.082660e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>SNP_17569439</td>\n",
       "      <td>3.014680e-17</td>\n",
       "      <td>3.014680e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SNP_17569017</td>\n",
       "      <td>3.014680e-17</td>\n",
       "      <td>3.014680e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>STR_17648335</td>\n",
       "      <td>4.115600e-17</td>\n",
       "      <td>4.115600e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>STR_17596478</td>\n",
       "      <td>1.510950e-17</td>\n",
       "      <td>1.510950e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>STR_17520209</td>\n",
       "      <td>6.019110e-18</td>\n",
       "      <td>6.019110e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>STR_17654342</td>\n",
       "      <td>4.725030e-18</td>\n",
       "      <td>4.725030e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>STR_17665017</td>\n",
       "      <td>3.137970e-18</td>\n",
       "      <td>3.137970e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>STR_17563147</td>\n",
       "      <td>3.084400e-18</td>\n",
       "      <td>3.084400e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>STR_17681081</td>\n",
       "      <td>2.948750e-18</td>\n",
       "      <td>2.948750e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>STR_17673720</td>\n",
       "      <td>2.836180e-18</td>\n",
       "      <td>2.836180e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>STR_17548440</td>\n",
       "      <td>2.127530e-18</td>\n",
       "      <td>2.127530e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>STR_17563928</td>\n",
       "      <td>1.917120e-18</td>\n",
       "      <td>1.917120e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>STR_17567263</td>\n",
       "      <td>1.657720e-18</td>\n",
       "      <td>1.657720e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>STR_17572568</td>\n",
       "      <td>1.642590e-18</td>\n",
       "      <td>1.642590e-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                0             1             2\n",
       "0    SNP_17586609  7.649090e-01  7.649090e-01\n",
       "1    SNP_17586715  1.064020e-01  1.064020e-01\n",
       "2    SNP_17586631  1.048040e-01  1.048040e-01\n",
       "3    SNP_17586583  2.388540e-02  2.388540e-02\n",
       "4    SNP_17602164  7.068190e-11  7.068190e-11\n",
       "5    SNP_17595915  4.266680e-11  4.266680e-11\n",
       "6    SNP_17595929  4.266680e-11  4.266680e-11\n",
       "7    SNP_17591089  2.184990e-11  2.184990e-11\n",
       "8    SNP_17596322  1.692930e-11  1.692930e-11\n",
       "9    SNP_17601466  5.988680e-12  5.988680e-12\n",
       "10   SNP_17582788  3.378290e-13  3.378290e-13\n",
       "11   SNP_17591144  2.538860e-13  2.538860e-13\n",
       "12   SNP_17585808  7.207200e-14  7.207200e-14\n",
       "13   SNP_17591697  4.957880e-14  4.957880e-14\n",
       "14   SNP_17605197  4.727070e-14  4.727070e-14\n",
       "15   SNP_17595070  2.404580e-14  2.404580e-14\n",
       "16   SNP_17594886  1.983560e-14  1.983560e-14\n",
       "17   SNP_17594915  1.983560e-14  1.983560e-14\n",
       "18   SNP_17582044  5.325070e-15  5.325070e-15\n",
       "19   SNP_17586471  2.582190e-15  2.582190e-15\n",
       "20   SNP_17587515  2.334290e-15  2.334290e-15\n",
       "21   SNP_17582061  2.159920e-15  2.159920e-15\n",
       "22   SNP_17582776  9.317800e-16  9.317800e-16\n",
       "23   SNP_17587746  9.254720e-16  9.254720e-16\n",
       "24   SNP_17583492  8.901370e-16  8.901370e-16\n",
       "25   SNP_17585465  8.301790e-16  8.301790e-16\n",
       "26   SNP_17599035  5.337240e-16  5.337240e-16\n",
       "27   SNP_17589152  5.171440e-16  5.171440e-16\n",
       "28   SNP_17606639  4.790680e-16  4.790680e-16\n",
       "29   SNP_17623675  4.737600e-16  4.737600e-16\n",
       "..            ...           ...           ...\n",
       "82   SNP_17566504  4.759770e-17  4.759770e-17\n",
       "83   SNP_17603801  4.700440e-17  4.700440e-17\n",
       "84   SNP_17584677  4.499790e-17  4.499790e-17\n",
       "85   SNP_17525584  4.443330e-17  4.443330e-17\n",
       "86   SNP_17598512  4.311950e-17  4.311950e-17\n",
       "87   SNP_17525970  4.120170e-17  4.120170e-17\n",
       "88   SNP_17589209  3.956490e-17  3.956490e-17\n",
       "89   SNP_17597056  3.644230e-17  3.644230e-17\n",
       "90   SNP_17609330  3.611320e-17  3.611320e-17\n",
       "91   SNP_17641500  3.582020e-17  3.582020e-17\n",
       "92   SNP_17668850  3.527790e-17  3.527790e-17\n",
       "93   SNP_17668851  3.527790e-17  3.527790e-17\n",
       "94   SNP_17567699  3.245120e-17  3.245120e-17\n",
       "95   SNP_17651273  3.121340e-17  3.121340e-17\n",
       "96   SNP_17650931  3.121340e-17  3.121340e-17\n",
       "97   SNP_17619378  3.082660e-17  3.082660e-17\n",
       "98   SNP_17569439  3.014680e-17  3.014680e-17\n",
       "99   SNP_17569017  3.014680e-17  3.014680e-17\n",
       "100  STR_17648335  4.115600e-17  4.115600e-17\n",
       "101  STR_17596478  1.510950e-17  1.510950e-17\n",
       "102  STR_17520209  6.019110e-18  6.019110e-18\n",
       "103  STR_17654342  4.725030e-18  4.725030e-18\n",
       "104  STR_17665017  3.137970e-18  3.137970e-18\n",
       "105  STR_17563147  3.084400e-18  3.084400e-18\n",
       "106  STR_17681081  2.948750e-18  2.948750e-18\n",
       "107  STR_17673720  2.836180e-18  2.836180e-18\n",
       "108  STR_17548440  2.127530e-18  2.127530e-18\n",
       "109  STR_17563928  1.917120e-18  1.917120e-18\n",
       "110  STR_17567263  1.657720e-18  1.657720e-18\n",
       "111  STR_17572568  1.642590e-18  1.642590e-18\n",
       "\n",
       "[112 rows x 3 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "post"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SNP_17586609' 0.7649090000000001 0.7649090000000001] \n",
      " ['SNP_17586631' 0.104804 0.104804] \n",
      " 4.1156000000000005e-17\n"
     ]
    }
   ],
   "source": [
    "            topvariants =  post.sort_values(post.columns[2], ascending=False).values[:5][0]\n",
    "            topscores  =  post.sort_values(post.columns[2], ascending=False).values[:5][2]\n",
    "            topcaviarstr  =  post.loc[post[0]=='STR_'+str(best_str_start)][2].tolist()[0]\n",
    "            print(topvariants, '\\n', topscores, '\\n', topcaviarstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
