{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adipose-Visceral/SNP_Analysis  variants ...\n",
      "(9113592, 15)   to   (16287, 15)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:58: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16287  total tests...  9926  pvalues were reduced to 1\n",
      "2078 \t gene level qval<=0.1\n",
      "471553 \t locus level qval<=0.1\n",
      "208872 \t qval<0.01\n",
      "\n",
      "Artery-Aorta/SNP_Analysis  variants ...\n",
      "(8889565, 15)   to   (16054, 15)\n",
      "16054  total tests...  9127  pvalues were reduced to 1\n",
      "2638 \t gene level qval<=0.1\n",
      "621312 \t locus level qval<=0.1\n",
      "276969 \t qval<0.01\n",
      "\n",
      "Artery-Tibial/SNP_Analysis  variants ...\n",
      "(8855753, 15)   to   (15774, 15)\n",
      "15774  total tests...  8014  pvalues were reduced to 1\n",
      "3728 \t gene level qval<=0.1\n",
      "888290 \t locus level qval<=0.1\n",
      "421953 \t qval<0.01\n",
      "\n",
      "Brain-Caudate/SNP_Analysis  variants ...\n",
      "(8413929, 15)   to   (16344, 15)\n",
      "16344  total tests...  11466  pvalues were reduced to 1\n",
      "761 \t gene level qval<=0.1\n",
      "153057 \t locus level qval<=0.1\n",
      "50316 \t qval<0.01\n",
      "\n",
      "Brain-Cerebellum/SNP_Analysis  variants ...\n",
      "(8385476, 15)   to   (16286, 15)\n",
      "16286  total tests...  10274  pvalues were reduced to 1\n",
      "1691 \t gene level qval<=0.1\n",
      "340146 \t locus level qval<=0.1\n",
      "128357 \t qval<0.01\n",
      "\n",
      "Cells-Transformedfibroblasts/SNP_Analysis  variants ...\n",
      "(8186925, 15)   to   (15354, 15)\n",
      "15354  total tests...  7968  pvalues were reduced to 1\n",
      "3326 \t gene level qval<=0.1\n",
      "703403 \t locus level qval<=0.1\n",
      "316458 \t qval<0.01\n",
      "\n",
      "Esophagus-Mucosa/SNP_Analysis  variants ...\n",
      "(8673113, 15)   to   (16146, 15)\n",
      "16146  total tests...  8308  pvalues were reduced to 1\n",
      "3664 \t gene level qval<=0.1\n",
      "790234 \t locus level qval<=0.1\n",
      "365810 \t qval<0.01\n",
      "\n",
      "Esophagus-Muscularis/SNP_Analysis  variants ...\n",
      "(8622045, 15)   to   (16068, 15)\n",
      "16068  total tests...  8834  pvalues were reduced to 1\n",
      "2975 \t gene level qval<=0.1\n",
      "645321 \t locus level qval<=0.1\n",
      "295152 \t qval<0.01\n",
      "\n",
      "Heart-LeftVentricle/SNP_Analysis  variants ...\n",
      "(8291644, 15)   to   (15558, 15)\n",
      "15558  total tests...  9588  pvalues were reduced to 1\n",
      "1883 \t gene level qval<=0.1\n",
      "401952 \t locus level qval<=0.1\n",
      "172585 \t qval<0.01\n",
      "\n",
      "Lung/SNP_Analysis  variants ...\n",
      "(8831998, 15)   to   (16415, 15)\n",
      "16415  total tests...  8907  pvalues were reduced to 1\n",
      "3167 \t gene level qval<=0.1\n",
      "682970 \t locus level qval<=0.1\n",
      "312850 \t qval<0.01\n",
      "\n",
      "Muscle-Skeletal/SNP_Analysis  variants ...\n",
      "(8360959, 15)   to   (15455, 15)\n",
      "15455  total tests...  8102  pvalues were reduced to 1\n",
      "3454 \t gene level qval<=0.1\n",
      "779541 \t locus level qval<=0.1\n",
      "380282 \t qval<0.01\n",
      "\n",
      "Nerve-Tibial/SNP_Analysis  variants ...\n",
      "(8843380, 15)   to   (16257, 15)\n",
      "16257  total tests...  7480  pvalues were reduced to 1\n",
      "4678 \t gene level qval<=0.1\n",
      "1027017 \t locus level qval<=0.1\n",
      "499121 \t qval<0.01\n",
      "\n",
      "Skin-NotSunExposed/SNP_Analysis  variants ...\n",
      "(8826938, 15)   to   (16381, 15)\n",
      "16381  total tests...  9392  pvalues were reduced to 1\n",
      "2557 \t gene level qval<=0.1\n",
      "536743 \t locus level qval<=0.1\n",
      "235991 \t qval<0.01\n",
      "\n",
      "Skin-SunExposed/SNP_Analysis  variants ...\n",
      "(8811928, 15)   to   (16306, 15)\n",
      "16306  total tests...  8018  pvalues were reduced to 1\n",
      "4095 \t gene level qval<=0.1\n",
      "900147 \t locus level qval<=0.1\n",
      "431375 \t qval<0.01\n",
      "\n",
      "Thyroid/SNP_Analysis  variants ...\n",
      "(8850459, 15)   to   (16322, 15)\n",
      "16322  total tests...  7793  pvalues were reduced to 1\n",
      "4281 \t gene level qval<=0.1\n",
      "961586 \t locus level qval<=0.1\n",
      "468886 \t qval<0.01\n",
      "\n",
      "WholeBlood/SNP_Analysis  variants ...\n",
      "(8010219, 15)   to   (14964, 15)\n",
      "14964  total tests...  8370  pvalues were reduced to 1\n",
      "2823 \t gene level qval<=0.1\n",
      "580804 \t locus level qval<=0.1\n",
      "282784 \t qval<0.01\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import collections\n",
    "import os\n",
    "\n",
    "## FDR correction After gene level adjustment\n",
    "SHORTEN = {\n",
    "    \"Artery-Aorta\":\"Artery A.\"     ,\n",
    "    \"Artery-Tibial\": \"Artery T.\",\n",
    "    \"Adipose-Subcutaneous\": \"Adipose S.\",    \n",
    "    \"Adipose-Visceral\":\"Adipose V.\",\n",
    "    \"Brain-Caudate\":\"Caudate\"   , \n",
    "    \"Brain-Cerebellum\":\"Cerebellum\",\n",
    "    \"Cells-Transformedfibroblasts\": \"Fibroblast\",\n",
    "    \"Esophagus-Mucosa\": \"E. Mucosa\",\n",
    "    \"Esophagus-Muscularis\":\"E Muscularis\",\n",
    "    \"Heart-LeftVentricle\":\"Ventricule\",\n",
    "    \"Lung\": \"Lung\",\n",
    "    \"Muscle-Skeletal\": \"Muscle\",\n",
    "    \"Nerve-Tibial\":\"Nerve\",\n",
    "    \"Skin-NotSunExposed\": \"Skin Unexposed\",\n",
    "    \"Skin-SunExposed\":\"Skin Leg\",\n",
    "    \"Thyroid\":\"Thyroid\",\n",
    "    \"WholeBlood\": \"Blood\"\n",
    "}\n",
    "\n",
    "tissu = sorted(SHORTEN.keys())\n",
    "#path = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/\"\n",
    "path = \"/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/\"\n",
    "\n",
    "def fdrcorrection(tissue):\n",
    "    print(tissue, ' variants ...')\n",
    "#Get most signif. variant by gene from linear reg STRs\n",
    "    LR1=pd.read_csv(path+ tissue+\"/Lin_Reg_Out\" , '\\t')\n",
    "    \n",
    "#Locus level\n",
    "    LR1['p.wald'].to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "    Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/fdr-correct.r\")\n",
    "    Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/qvalues.txt', sep=' ')\n",
    "    \n",
    "    LR1['llqvalue']=list(Qval['qvalue'])\n",
    "    LR1['llsignif']=list(Qval['significant'])\n",
    "\n",
    "#Gene level\n",
    "    LR0 = LR1.sort_values(\"p.wald\").groupby(\"gene\", as_index=False).first()     \n",
    "    print(LR1.shape, '  to  ', LR0.shape)\n",
    "\n",
    "    #Add counts tests by gene\n",
    "    counts=pd.DataFrame({'cts' : LR1.groupby([\"gene\"]).size()})    ## This is the count by genes\n",
    "    genes = list(LR0['gene'])\n",
    "    LR0['NTEST']= list(counts.loc[genes]['cts'])\n",
    "    \n",
    "    #Gene level adjustment\n",
    "    #(1) min_pval* #test\n",
    "    LR0['AD.pval']=LR0['p.wald']*LR0['NTEST']\n",
    "    #(2) if AD_pval>1 => AD_pval=1\n",
    "    LR0['AD.pval'][LR0['AD.pval']>1] = 1\n",
    "    \n",
    "    #Save pval in file and FDR correct\n",
    "    LR0['AD.pval'].to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "    Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/fdr-correct.r\")\n",
    "        \n",
    "    #FDR corrected... add to dataframe\n",
    "    Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/qvalues.txt', sep=' ')\n",
    "    LR0['qvalue']=list(Qval['qvalue'])\n",
    "    LR0['significant']=list(Qval['significant'])\n",
    "\n",
    "#Merging\n",
    "    merging=['gene','chrom','str.id','str.start','beta','beta.se','p.wald','llqvalue','llsignif']\n",
    "    LRP = pd.merge(LR1,LR0, on=merging, how='left')\n",
    "    \n",
    "#Header arrangement\n",
    "    Head=['gene','chrom','str.id','str.start','p.wald','llqvalue','llsignif','NTEST','qvalue','significant','beta','beta.se']\n",
    "    Out=LRP[Head]\n",
    "    Out.to_csv(path+tissue+'/PQValues', sep='\\t', index=False)\n",
    "\n",
    "    S=LR0['AD.pval']\n",
    "    print(len(S),' total tests... ', len(S[S>=1]) , ' pvalues were reduced to 1')\n",
    "    print(len(LRP[LRP['qvalue'] <=0.1]),'\\t gene level qval<=0.1')\n",
    "    print(len(LRP[LRP['llqvalue'] <=0.1]),'\\t locus level qval<=0.1')\n",
    "    print(len(LRP[LRP['llqvalue'] <=0.01]),'\\t qval<0.01\\n')\n",
    "    return()\n",
    "#\n",
    "#\n",
    "for T in tissu[1:]:                     \n",
    "    fdrcorrection(T+'/SNP_Analysis')\n",
    "    \n",
    "#fdrcorrection('Esophagus-Mucosa')\n",
    "####   After this Script, plot the qqplot and then FDR correction below    \n",
    "####   Next: We run the code to prepare for heritability analysis (STR+SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WholeBlood  variants ...\n",
      "(209131, 12)   to   (14845, 12)\n",
      "14845\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14845  total tests...  5015  pvalues were reduced to 1\n",
      "\n",
      "630 \t qval<=0.1\n",
      "462 \t qval<=0.05\n",
      "261 \t qval<0.01\n",
      "\n",
      "Cells-Transformedfibroblasts  variants ...\n",
      "(216353, 12)   to   (15362, 12)\n",
      "15362\n",
      "15362  total tests...  4885  pvalues were reduced to 1\n",
      "\n",
      "955 \t qval<=0.1\n",
      "703 \t qval<=0.05\n",
      "395 \t qval<0.01\n",
      "\n",
      "Muscle-Skeletal  variants ...\n",
      "(215860, 12)   to   (15275, 12)\n",
      "15275\n",
      "15275  total tests...  5546  pvalues were reduced to 1\n",
      "\n",
      "304 \t qval<=0.1\n",
      "205 \t qval<=0.05\n",
      "0 \t qval<0.01\n",
      "\n",
      "Lung  variants ...\n",
      "(225784, 12)   to   (15711, 12)\n",
      "15711\n",
      "15711  total tests...  5673  pvalues were reduced to 1\n",
      "\n",
      "378 \t qval<=0.1\n",
      "264 \t qval<=0.05\n",
      "0 \t qval<0.01\n",
      "\n",
      "Adipose-Subcutaneous  variants ...\n",
      "(225770, 12)   to   (15918, 12)\n",
      "15918\n",
      "15918  total tests...  5751  pvalues were reduced to 1\n",
      "\n",
      "376 \t qval<=0.1\n",
      "287 \t qval<=0.05\n",
      "115 \t qval<0.01\n",
      "\n",
      "Artery-Tibial  variants ...\n",
      "(221263, 12)   to   (15584, 12)\n",
      "15584\n",
      "15584  total tests...  5360  pvalues were reduced to 1\n",
      "\n",
      "524 \t qval<=0.1\n",
      "383 \t qval<=0.05\n",
      "196 \t qval<0.01\n",
      "\n",
      "Esophagus-Mucosa  variants ...\n",
      "(227378, 12)   to   (16111, 12)\n",
      "16111\n",
      "16111  total tests...  5582  pvalues were reduced to 1\n",
      "\n",
      "475 \t qval<=0.1\n",
      "330 \t qval<=0.05\n",
      "128 \t qval<0.01\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
