{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nerve-Tibial/SNP_Analysis  variants ...\n",
      "(8843380, 13)   to   (16257, 13)\n",
      "16257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:48: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16257  total tests...  7480  pvalues were reduced to 1\n",
      "4678 \t qval<=0.1\n",
      "4178 \t qval<=0.05\n",
      "3455 \t qval<0.01\n",
      "\n",
      "Muscle-Skeletal/SNP_Analysis  variants ...\n",
      "(8360959, 13)   to   (15455, 13)\n",
      "15455\n",
      "15455  total tests...  8102  pvalues were reduced to 1\n",
      "3454 \t qval<=0.1\n",
      "3105 \t qval<=0.05\n",
      "2448 \t qval<0.01\n",
      "\n",
      "Skin-SunExposed(Lowerleg)/SNP_Analysis  variants ...\n",
      "(8811928, 13)   to   (16306, 13)\n",
      "16306\n",
      "16306  total tests...  8018  pvalues were reduced to 1\n",
      "4095 \t qval<=0.1\n",
      "3645 \t qval<=0.05\n",
      "2886 \t qval<0.01\n",
      "\n",
      "Cells-Transformedfibroblasts/SNP_Analysis  variants ...\n",
      "(8186925, 13)   to   (15354, 13)\n",
      "15354\n",
      "15354  total tests...  7968  pvalues were reduced to 1\n",
      "3326 \t qval<=0.1\n",
      "2993 \t qval<=0.05\n",
      "2350 \t qval<0.01\n",
      "\n",
      "Brain-Caudate(basalganglia)/SNP_Analysis  variants ...\n",
      "(8413929, 13)   to   (16344, 13)\n",
      "16344\n",
      "16344  total tests...  11466  pvalues were reduced to 1\n",
      "761 \t qval<=0.1\n",
      "635 \t qval<=0.05\n",
      "398 \t qval<0.01\n",
      "\n",
      "Esophagus-Muscularis/SNP_Analysis  variants ...\n",
      "(8622045, 13)   to   (16068, 13)\n",
      "16068\n",
      "16068  total tests...  8834  pvalues were reduced to 1\n",
      "2975 \t qval<=0.1\n",
      "2615 \t qval<=0.05\n",
      "2062 \t qval<0.01\n",
      "\n",
      "Skin-NotSunExposed(Suprapubic)/SNP_Analysis  variants ...\n",
      "(8826938, 13)   to   (16381, 13)\n",
      "16381\n",
      "16381  total tests...  9392  pvalues were reduced to 1\n",
      "2557 \t qval<=0.1\n",
      "2207 \t qval<=0.05\n",
      "1650 \t qval<0.01\n",
      "\n",
      "Heart-LeftVentricle/SNP_Analysis  variants ...\n",
      "(8291644, 13)   to   (15558, 13)\n",
      "15558\n",
      "15558  total tests...  9588  pvalues were reduced to 1\n",
      "1883 \t qval<=0.1\n",
      "1633 \t qval<=0.05\n",
      "1241 \t qval<0.01\n",
      "\n",
      "Thyroid/SNP_Analysis  variants ...\n",
      "(8850459, 13)   to   (16322, 13)\n",
      "16322\n",
      "16322  total tests...  7793  pvalues were reduced to 1\n",
      "4281 \t qval<=0.1\n",
      "3838 \t qval<=0.05\n",
      "3130 \t qval<0.01\n",
      "\n",
      "Artery-Tibial/SNP_Analysis  variants ...\n",
      "(8855753, 13)   to   (15774, 13)\n",
      "15774\n",
      "15774  total tests...  8014  pvalues were reduced to 1\n",
      "3728 \t qval<=0.1\n",
      "3364 \t qval<=0.05\n",
      "2695 \t qval<0.01\n",
      "\n",
      "Esophagus-Mucosa/SNP_Analysis  variants ...\n",
      "(8673113, 13)   to   (16146, 13)\n",
      "16146\n",
      "16146  total tests...  8308  pvalues were reduced to 1\n",
      "3664 \t qval<=0.1\n",
      "3231 \t qval<=0.05\n",
      "2510 \t qval<0.01\n",
      "\n",
      "WholeBlood/SNP_Analysis  variants ...\n",
      "(8010219, 13)   to   (14964, 13)\n",
      "14964\n",
      "14964  total tests...  8370  pvalues were reduced to 1\n",
      "2823 \t qval<=0.1\n",
      "2538 \t qval<=0.05\n",
      "2051 \t qval<0.01\n",
      "\n",
      "Lung/SNP_Analysis  variants ...\n",
      "(8831998, 13)   to   (16415, 13)\n",
      "16415\n",
      "16415  total tests...  8907  pvalues were reduced to 1\n",
      "3167 \t qval<=0.1\n",
      "2815 \t qval<=0.05\n",
      "2207 \t qval<0.01\n",
      "\n",
      "Adipose-Subcutaneous/SNP_Analysis  variants ...\n",
      "(9069324, 13)   to   (16155, 13)\n",
      "16155\n",
      "16155  total tests...  8147  pvalues were reduced to 1\n",
      "3938 \t qval<=0.1\n",
      "3524 \t qval<=0.05\n",
      "2808 \t qval<0.01\n",
      "\n",
      "Brain-Cerebellum/SNP_Analysis  variants ...\n",
      "(8385476, 13)   to   (16286, 13)\n",
      "16286\n",
      "16286  total tests...  10274  pvalues were reduced to 1\n",
      "1691 \t qval<=0.1\n",
      "1425 \t qval<=0.05\n",
      "1011 \t qval<0.01\n",
      "\n",
      "Adipose-Visceral(Omentum)/SNP_Analysis  variants ...\n",
      "(9113592, 13)   to   (16287, 13)\n",
      "16287\n",
      "16287  total tests...  9926  pvalues were reduced to 1\n",
      "2078 \t qval<=0.1\n",
      "1809 \t qval<=0.05\n",
      "1392 \t qval<0.01\n",
      "\n",
      "Artery-Aorta/SNP_Analysis  variants ...\n",
      "(8889565, 13)   to   (16054, 13)\n",
      "16054\n",
      "16054  total tests...  9127  pvalues were reduced to 1\n",
      "2638 \t qval<=0.1\n",
      "2327 \t qval<=0.05\n",
      "1810 \t qval<0.01\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## FDR correction After gene level adjustment\n",
    "SHORTEN = {\n",
    "    \"Artery-Aorta\":\"Artery A.\"     ,\n",
    "    \"Artery-Tibial\": \"Artery T.\",\n",
    "    \"Adipose-Subcutaneous\": \"Adipose S.\",    \n",
    "    \"Adipose-Visceral(Omentum)\":\"Adipose V.\",\n",
    "    \"Brain-Caudate(basalganglia)\":\"Caudate\"   , \n",
    "    \"Brain-Cerebellum\":\"Cerebellum\",\n",
    "    \"Cells-Transformedfibroblasts\": \"Fibroblast\",\n",
    "    \"Esophagus-Mucosa\": \"E. Mucosa\",\n",
    "    \"Esophagus-Muscularis\":\"E Muscularis\",\n",
    "    \"Heart-LeftVentricle\":\"Ventricule\",\n",
    "    \"Lung\": \"Lung\",\n",
    "    \"Muscle-Skeletal\": \"Muscle\",\n",
    "    \"Nerve-Tibial\":\"Nerve\",\n",
    "    \"Skin-NotSunExposed(Suprapubic)\": \"Skin Unexposed\",\n",
    "    \"Skin-SunExposed(Lowerleg)\":\"Skin Leg\",\n",
    "    \"Thyroid\":\"Thyroid\",\n",
    "    \"WholeBlood\": \"Blood\"\n",
    "}\n",
    "\n",
    "tissu = SHORTEN.keys()\n",
    "#path = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/\"\n",
    "path = \"/storage/szfeupe/Runs/650GTEx_estr/Analysis_by_Tissue/\"\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import collections\n",
    "def fdrcorrection(tissue):\n",
    "    print(tissue, ' variants ...')\n",
    "#Get most signif. variant by gene from linear reg STRs\n",
    "    LR1=pd.read_csv(path+ tissue+\"/Lin_Reg_Out\" , '\\t')       #/Lin_Reg_OutFin.txt\n",
    "    LR0 = LR1.sort_values(\"p.wald\").groupby(\"gene\", as_index=False).first()     \n",
    "    print(LR1.shape, '  to  ', LR0.shape)\n",
    "\n",
    "#Add counts tests by gene\n",
    "    counts=pd.DataFrame({'cts' : LR1.groupby([\"gene\"]).size()})    ## This is the count by genes\n",
    "    genes = list(LR0['gene'])\n",
    "    LR0['NTEST']= list(counts.loc[genes]['cts'])\n",
    "    print (len(counts.loc[genes]['cts']))\n",
    "    \n",
    "#Minor adjustment\n",
    "    #(1) min_pval* #test\n",
    "    LR0['AD.pval']=LR0['p.wald']*LR0['NTEST']\n",
    "    #(2) if AD_pval>1 => AD_pval=1\n",
    "    LR0['AD.pval'][LR0['AD.pval']>1] = 1\n",
    "    \n",
    "#Save pval in file and FDR correct\n",
    "    LR0['AD.pval'].to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "    Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/fdr-correct.r\")\n",
    "    \n",
    "#FDR corrected... add to dataframe\n",
    "    Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/qvalues.txt', sep=' ')\n",
    "    #print (Qval.shape, '\\t', LR0.shape)\n",
    "    LR0['qvalue']=list(Qval['qvalue'])\n",
    "    LR0['significant']=list(Qval['significant'])\n",
    "    \n",
    "#Header arrangement\n",
    "    Head=['gene','chrom','str.id','str.start','NTEST','p.wald','AD.pval','qvalue','significant','beta','beta.se']\n",
    "    Out=LR0[Head]\n",
    "    Out.to_csv(path+tissue+'/PQValues', sep='\\t', index=False)\n",
    "\n",
    "    S=LR0['AD.pval']\n",
    "    print(len(S),' total tests... ', len(S[S>=1]) , ' pvalues were reduced to 1')\n",
    "    print(len(Qval[Qval['qvalue'] <=0.1]),'\\t qval<=0.1')\n",
    "    print(len(Qval[Qval['qvalue'] <=0.05]),'\\t qval<=0.05')\n",
    "    print(len(Qval[Qval['qvalue'] <=0.01]),'\\t qval<0.01\\n')\n",
    "#   print('Original linear reg. tests counts\\n',LR1.groupby('chrom').size())\n",
    "#   print('After FDR correction test counts\\n',LR0.groupby('chrom').size())    \n",
    "    return()\n",
    "\n",
    "#fdrcorrection('Lung/SNP_Analysis')\n",
    "for T in tissu:                     #[x for x in tissu if 'Brain' not in x and x not in ['WholeBlood','Esophagus-Mucosa','Skin-SunExposed(Lowerleg)','Heart-LeftVentricle','Artery-Tibial','Muscle-Skeletal','Adipose-Visceral(Omentum)','Artery-Aorta','Esophagus-Muscularis','Adipose-Subcutaneous','Cells-Transformedfibroblasts','Thyroid']]:                #tissu:\n",
    "    fdrcorrection(T+'/SNP_Analysis')\n",
    "    \n",
    "##########   Next: We run the code to prepare for heritability analysis (STR+SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Tissues done\n",
    "\"Adipose-Subcutaneous\",\"Artery-Tibial\"\n",
    "\"Adipose-Visceral(Omentum)\",\"Artery-Aorta\"\n",
    "\"Cells-Transformedfibroblasts\",\"WholeBlood\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WholeBlood/SNP_Analysis\n",
      "(902066, 13) (902035, 13) (901976, 13) End\n",
      "(630028, 13) (629982, 13) (629958, 13) End\n",
      "(577360, 13) (577338, 13) (577302, 13) End\n",
      "(429113, 13) (429097, 13) (429061, 13) End\n",
      "(438934, 13) (438903, 13) (438879, 13) End\n",
      "(637134, 13) (637120, 13) (637075, 13) End\n",
      "(487894, 13) (487871, 13) (487852, 13) End\n",
      "(371171, 13) (371163, 13) (371144, 13) End\n",
      "(382213, 13) (382195, 13) (382175, 13) End\n",
      "(422058, 13) (422046, 13) (422011, 13) End\n",
      "(508536, 13) (508522, 13) (508486, 13) End\n",
      "(503939, 13) (503926, 13) (503908, 13) End\n",
      "(186898, 13) (186897, 13) (186890, 13) End\n",
      "(309828, 13) (309798, 13) (309779, 13) End\n",
      "(297667, 13) (297655, 13) (297637, 13) End\n",
      "(413588, 13) (413585, 13) (413567, 13) End\n",
      "(486064, 13) (486019, 13) (486002, 13) End\n",
      "(173099, 13) (173095, 13) (173088, 13) End\n",
      "(697074, 13) (697060, 13) (697032, 13) End\n",
      "(263576, 13) (263552, 13) (263532, 13) End\n",
      "(117778, 13) (117774, 13) (117772, 13) End\n",
      "(238224, 13) (238204, 13) (238199, 13) End\n",
      "(176410, 13) (176399, 13) (176383, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "Cells-Transformedfibroblasts/SNP_Analysis\n",
      "(919571, 13) (919539, 13) (919478, 13) End\n",
      "(668690, 13) (668644, 13) (668616, 13) End\n",
      "(595685, 13) (595663, 13) (595628, 13) End\n",
      "(456948, 13) (456927, 13) (456888, 13) End\n",
      "(475851, 13) (475805, 13) (475782, 13) End\n",
      "(652426, 13) (652412, 13) (652369, 13) End\n",
      "(511925, 13) (511902, 13) (511888, 13) End\n",
      "(395173, 13) (395163, 13) (395142, 13) End\n",
      "(407115, 13) (407097, 13) (407076, 13) End\n",
      "(417027, 13) (417014, 13) (416978, 13) End\n",
      "(531433, 13) (531418, 13) (531384, 13) End\n",
      "(523343, 13) (523331, 13) (523312, 13) End\n",
      "(204583, 13) (204582, 13) (204575, 13) End\n",
      "(313136, 13) (313106, 13) (313087, 13) End\n",
      "(298809, 13) (298798, 13) (298779, 13) End\n",
      "(416311, 13) (416306, 13) (416289, 13) End\n",
      "(501223, 13) (501176, 13) (501159, 13) End\n",
      "(183506, 13) (183502, 13) (183494, 13) End\n",
      "(680310, 13) (680296, 13) (680268, 13) End\n",
      "(257398, 13) (257374, 13) (257354, 13) End\n",
      "(122633, 13) (122629, 13) (122628, 13) End\n",
      "(233244, 13) (233223, 13) (233217, 13) End\n",
      "(190217, 13) (190206, 13) (190190, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "Muscle-Skeletal/SNP_Analysis\n",
      "(930748, 13) (930715, 13) (930657, 13) End\n",
      "(670758, 13) (670714, 13) (670687, 13) End\n",
      "(600238, 13) (600216, 13) (600181, 13) End\n",
      "(440513, 13) (440497, 13) (440459, 13) End\n",
      "(472895, 13) (472851, 13) (472828, 13) End\n",
      "(649914, 13) (649900, 13) (649857, 13) End\n",
      "(507184, 13) (507161, 13) (507146, 13) End\n",
      "(392948, 13) (392938, 13) (392917, 13) End\n",
      "(399757, 13) (399737, 13) (399716, 13) End\n",
      "(428954, 13) (428941, 13) (428904, 13) End\n",
      "(530198, 13) (530183, 13) (530146, 13) End\n",
      "(515731, 13) (515719, 13) (515700, 13) End\n",
      "(196416, 13) (196415, 13) (196408, 13) End\n",
      "(319876, 13) (319845, 13) (319826, 13) End\n",
      "(304081, 13) (304071, 13) (304053, 13) End\n",
      "(423640, 13) (423635, 13) (423616, 13) End\n",
      "(495602, 13) (495555, 13) (495538, 13) End\n",
      "(175714, 13) (175710, 13) (175703, 13) End\n",
      "(680983, 13) (680969, 13) (680943, 13) End\n",
      "(261524, 13) (261499, 13) (261478, 13) End\n",
      "(121821, 13) (121817, 13) (121815, 13) End\n",
      "(236508, 13) (236488, 13) (236482, 13) End\n",
      "(193247, 13) (193236, 13) (193220, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "Lung/SNP_Analysis\n",
      "(986731, 13) (986697, 13) (986633, 13) End\n",
      "(707505, 13) (707458, 13) (707427, 13) End\n",
      "(632624, 13) (632602, 13) (632565, 13) End\n",
      "(487161, 13) (487140, 13) (487099, 13) End\n",
      "(509889, 13) (509842, 13) (509817, 13) End\n",
      "(700648, 13) (700633, 13) (700587, 13) End\n",
      "(538963, 13) (538939, 13) (538923, 13) End\n",
      "(421845, 13) (421833, 13) (421811, 13) End\n",
      "(420563, 13) (420544, 13) (420522, 13) End\n",
      "(475823, 13) (475810, 13) (475769, 13) End\n",
      "(573067, 13) (573052, 13) (573019, 13) End\n",
      "(562766, 13) (562753, 13) (562733, 13) End\n",
      "(213197, 13) (213196, 13) (213188, 13) End\n",
      "(330879, 13) (330849, 13) (330830, 13) End\n",
      "(323676, 13) (323665, 13) (323645, 13) End\n",
      "(450617, 13) (450612, 13) (450593, 13) End\n",
      "(522982, 13) (522935, 13) (522918, 13) End\n",
      "(190658, 13) (190654, 13) (190649, 13) End\n",
      "(738764, 13) (738750, 13) (738720, 13) End\n",
      "(285210, 13) (285183, 13) (285160, 13) End\n",
      "(134506, 13) (134502, 13) (134499, 13) End\n",
      "(252832, 13) (252811, 13) (252805, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "Adipose-Subcutaneous/SNP_Analysis\n",
      "(963594, 13) (963560, 13) (963499, 13) End\n",
      "(699718, 13) (699672, 13) (699641, 13) End\n",
      "(614722, 13) (614700, 13) (614663, 13) End\n",
      "(473555, 13) (473534, 13) (473494, 13) End\n",
      "(495334, 13) (495287, 13) (495263, 13) End\n",
      "(674633, 13) (674618, 13) (674575, 13) End\n",
      "(527276, 13) (527252, 13) (527237, 13) End\n",
      "(414752, 13) (414741, 13) (414726, 13) End\n",
      "(414044, 13) (414025, 13) (414003, 13) End\n",
      "(463033, 13) (463020, 13) (462980, 13) End\n",
      "(555313, 13) (555296, 13) (555265, 13) End\n",
      "(544375, 13) (544363, 13) (544344, 13) End\n",
      "(209346, 13) (209345, 13) (209336, 13) End\n",
      "(326292, 13) (326262, 13) (326242, 13) End\n",
      "(314004, 13) (313993, 13) (313974, 13) End\n",
      "(440210, 13) (440205, 13) (440186, 13) End\n",
      "(511837, 13) (511798, 13) (511781, 13) End\n",
      "(186963, 13) (186959, 13) (186953, 13) End\n",
      "(721926, 13) (721912, 13) (721890, 13) End\n",
      "(269637, 13) (269613, 13) (269591, 13) End\n",
      "(125919, 13) (125915, 13) (125914, 13) End\n",
      "(241443, 13) (241423, 13) (241417, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "Artery-Tibial/SNP_Analysis\n",
      "(946463, 13) (946429, 13) (946369, 13) End\n",
      "(691670, 13) (691624, 13) (691591, 13) End\n",
      "(615714, 13) (615693, 13) (615656, 13) End\n",
      "(460897, 13) (460878, 13) (460839, 13) End\n",
      "(491307, 13) (491260, 13) (491236, 13) End\n",
      "(662739, 13) (662723, 13) (662678, 13) End\n",
      "(518286, 13) (518264, 13) (518245, 13) End\n",
      "(396403, 13) (396393, 13) (396372, 13) End\n",
      "(404724, 13) (404704, 13) (404682, 13) End\n",
      "(443334, 13) (443321, 13) (443282, 13) End\n",
      "(554380, 13) (554364, 13) (554328, 13) End\n",
      "(531281, 13) (531269, 13) (531250, 13) End\n",
      "(207995, 13) (207994, 13) (207986, 13) End\n",
      "(321201, 13) (321171, 13) (321152, 13) End\n",
      "(307270, 13) (307259, 13) (307240, 13) End\n",
      "(428736, 13) (428731, 13) (428713, 13) End\n",
      "(501594, 13) (501548, 13) (501531, 13) End\n",
      "(183520, 13) (183516, 13) (183509, 13) End\n",
      "(691468, 13) (691454, 13) (691427, 13) End\n",
      "(265299, 13) (265276, 13) (265254, 13) End\n",
      "(129105, 13) (129101, 13) (129099, 13) End\n",
      "(239067, 13) (239047, 13) (239041, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "Esophagus-Mucosa/SNP_Analysis\n",
      "(981166, 13) (981133, 13) (981069, 13) End\n",
      "(705710, 13) (705662, 13) (705632, 13) End\n",
      "(616210, 13) (616188, 13) (616154, 13) End\n",
      "(479277, 13) (479257, 13) (479218, 13) End\n",
      "(498914, 13) (498867, 13) (498843, 13) End\n",
      "(690326, 13) (690311, 13) (690268, 13) End\n",
      "(532614, 13) (532592, 13) (532577, 13) End\n",
      "(404828, 13) (404816, 13) (404795, 13) End\n",
      "(412039, 13) (412021, 13) (412000, 13) End\n",
      "(463843, 13) (463830, 13) (463789, 13) End\n",
      "(566467, 13) (566451, 13) (566419, 13) End\n",
      "(549974, 13) (549962, 13) (549944, 13) End\n",
      "(211421, 13) (211420, 13) (211412, 13) End\n",
      "(326535, 13) (326505, 13) (326485, 13) End\n",
      "(315760, 13) (315748, 13) (315728, 13) End\n",
      "(447966, 13) (447961, 13) (447942, 13) End\n",
      "(523334, 13) (523294, 13) (523277, 13) End\n",
      "(192115, 13) (192111, 13) (192105, 13) End\n",
      "(730810, 13) (730795, 13) (730772, 13) End\n",
      "(275148, 13) (275122, 13) (275100, 13) End\n",
      "(130628, 13) (130624, 13) (130623, 13) End\n",
      "(246169, 13) (246148, 13) (246142, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n",
      "(0, 13) (0, 13) (0, 13) End\n"
     ]
    }
   ],
   "source": [
    "#           MAKE SURE YOU HAVE THE REGRESSION OUTPUT MERGED FIRST ie all chrs in 1 file\n",
    "tissu = ['WholeBlood','Cells-Transformedfibroblasts','Muscle-Skeletal','Lung','Adipose-Subcutaneous','Artery-Tibial','Esophagus-Mucosa']\n",
    "\n",
    "##This code will remove STRs that are homopolymer or fall in \n",
    "##least polymorphic sites from linear regression analysis output\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def filterregression(Tissu):\n",
    "    #\n",
    "    print(Tissu)\n",
    "    chrs = [i for i in range(1,23)]\n",
    "    chrs.append('X') ; chrs.append('Y')\n",
    "\n",
    "    #regression\n",
    "    Input='/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/'+Tissu\n",
    "    linreg = pd.read_csv(Input+'/Lin_Reg_Out', sep='\\t')\n",
    "#    linreg = pd.read_csv(Input+'/Lin_Reg_Out_perm', sep='\\t')\n",
    "\n",
    "    #Homopolymers\n",
    "    strfile= \"/storage/szfeupe/Runs/GTEx_estr/Homopolymer_sites.tsv\"\n",
    "    L_pol = pd.read_csv(strfile, sep='\\t')\n",
    "\n",
    "    index=['0']\n",
    "    Out = pd.DataFrame(index=index, columns=linreg.columns)\n",
    "\n",
    "    #LR Cleanup\n",
    "    for i in chrs:\n",
    "    #Open sites to be removed\n",
    "        Lpolym = '/storage/szfeupe/Runs/GTEx_estr/Heterozygosity/Heterozygosity_VCF.chr'+str(i)\n",
    "        Reg_to_remove=pd.read_csv(Lpolym, sep='\\t', header=None)\n",
    "        Reg_to_remove.columns=['CHROM','START','END','NumAlleleLenghts','NumAlleles','Het.length','Het.genotype']\n",
    "        Homopoly = L_pol.loc[L_pol['chrom']=='chr'+str(i)]\n",
    "\n",
    "    #   Remove Homoploymers\n",
    "        LIN = linreg.loc[linreg['chrom']=='chr'+str(i)]\n",
    "        LIN0 = LIN.loc[LIN['str.start'].isin(list(Homopoly['str.start']))==False]\n",
    "\n",
    "    #   Remove least polymorphic\n",
    "        Reg=Reg_to_remove.loc[Reg_to_remove['Het.length']<=0.3]\n",
    "        FIN = LIN0.loc[LIN0['str.start'].isin(list(Reg['START']))==False]\n",
    "\n",
    "    #   Sum it up and outout   \n",
    "        frames = [Out, FIN]\n",
    "        Out = pd.concat(frames)\n",
    "\n",
    "        print(LIN.shape, LIN0.shape, FIN.shape, 'End')\n",
    "\n",
    "    Out1=Out.drop('0')\n",
    "    Out1=Out1.drop('Unnamed: 0', 1)\n",
    "    Out1.to_csv(Input+'/Lin_Reg_OutFin.txt', sep='\\t', index=False)\n",
    "    #Out1.to_csv(Input+'/Lin_Reg_Out_perm_Fin.txt', sep='\\t', index=False)\n",
    "\n",
    "for T in tissu:\n",
    "    filterregression(T+'/SNP_Analysis')\n",
    "    \n",
    "####After this Script, plot the qqplot and then FDR correction below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WholeBlood  variants ...\n",
      "(209131, 12)   to   (14845, 12)\n",
      "14845\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14845  total tests...  5015  pvalues were reduced to 1\n",
      "\n",
      "630 \t qval<=0.1\n",
      "462 \t qval<=0.05\n",
      "261 \t qval<0.01\n",
      "\n",
      "Cells-Transformedfibroblasts  variants ...\n",
      "(216353, 12)   to   (15362, 12)\n",
      "15362\n",
      "15362  total tests...  4885  pvalues were reduced to 1\n",
      "\n",
      "955 \t qval<=0.1\n",
      "703 \t qval<=0.05\n",
      "395 \t qval<0.01\n",
      "\n",
      "Muscle-Skeletal  variants ...\n",
      "(215860, 12)   to   (15275, 12)\n",
      "15275\n",
      "15275  total tests...  5546  pvalues were reduced to 1\n",
      "\n",
      "304 \t qval<=0.1\n",
      "205 \t qval<=0.05\n",
      "0 \t qval<0.01\n",
      "\n",
      "Lung  variants ...\n",
      "(225784, 12)   to   (15711, 12)\n",
      "15711\n",
      "15711  total tests...  5673  pvalues were reduced to 1\n",
      "\n",
      "378 \t qval<=0.1\n",
      "264 \t qval<=0.05\n",
      "0 \t qval<0.01\n",
      "\n",
      "Adipose-Subcutaneous  variants ...\n",
      "(225770, 12)   to   (15918, 12)\n",
      "15918\n",
      "15918  total tests...  5751  pvalues were reduced to 1\n",
      "\n",
      "376 \t qval<=0.1\n",
      "287 \t qval<=0.05\n",
      "115 \t qval<0.01\n",
      "\n",
      "Artery-Tibial  variants ...\n",
      "(221263, 12)   to   (15584, 12)\n",
      "15584\n",
      "15584  total tests...  5360  pvalues were reduced to 1\n",
      "\n",
      "524 \t qval<=0.1\n",
      "383 \t qval<=0.05\n",
      "196 \t qval<0.01\n",
      "\n",
      "Esophagus-Mucosa  variants ...\n",
      "(227378, 12)   to   (16111, 12)\n",
      "16111\n",
      "16111  total tests...  5582  pvalues were reduced to 1\n",
      "\n",
      "475 \t qval<=0.1\n",
      "330 \t qval<=0.05\n",
      "128 \t qval<0.01\n",
      "\n"
     ]
    }
   ],
   "source": [
    "tissu = ['WholeBlood','Cells-Transformedfibroblasts','Muscle-Skeletal','Lung','Adipose-Subcutaneous','Artery-Tibial','Esophagus-Mucosa']\n",
    "\n",
    "##Simpler version of FDR correction \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import collections\n",
    "def fdrcorrection(tissue):\n",
    "    print(tissue, ' variants ...')\n",
    "#Get most signif. variant by gene from linear reg STRs\n",
    "    LR1=pd.read_csv(\"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/\"+ tissue+\"/Lin_Reg_OutFin.txt\" , '\\t')\n",
    "    LR0 = LR1.sort_values(\"p.wald\").groupby(\"gene\", as_index=False).first()     \n",
    "    print(LR1.shape, '  to  ', LR0.shape)\n",
    "\n",
    "#Add counts tests by gene\n",
    "    counts=pd.DataFrame({'cts' : LR1.groupby([\"gene\"]).size()})    ## This is the count by genes\n",
    "    genes = list(LR0['gene'])\n",
    "    LR0['NTEST']= list(counts.loc[genes]['cts'])\n",
    "    print (len(counts.loc[genes]['cts']))\n",
    "    \n",
    "#Minor adjustment\n",
    "    #(1) min_pval* #test\n",
    "    LR0['AD.pval']=LR0['p.wald']*LR0['NTEST']\n",
    "    #(2) if AD_pval>1 => AD_pval=1\n",
    "    LR0['AD.pval'][LR0['AD.pval']>1] = 1\n",
    "    \n",
    "#Save pval in file and FDR correct\n",
    "    LR0['AD.pval'].to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "    Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/fdr-correct.r\")\n",
    "    \n",
    "#FDR corrected... add to dataframe\n",
    "    Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/PostRegressionFixes/qvalues.txt', sep=' ')\n",
    "    #print (Qval.shape, '\\t', LR0.shape)\n",
    "    LR0['qvalue']=list(Qval['qvalue'])\n",
    "    LR0['significant']=list(Qval['significant'])\n",
    "    \n",
    "#Header arrangement\n",
    "    Head=['gene','chrom','str.id','str.start','NTEST','p.wald','AD.pval','qvalue','significant','beta','beta.se']\n",
    "    Out=LR0[Head]\n",
    "    Out.to_csv(\"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/\"+tissue+'/PQValues.txt', sep='\\t', index=False)\n",
    "\n",
    "    S=LR0['AD.pval']\n",
    "    print(len(S),' total tests... ', len(S[S>=1]) , ' pvalues were reduced to 1')\n",
    "    print(len(Qval[Qval['qvalue'] <=0.1]),'\\t qval<=0.1')\n",
    "    print(len(Qval[Qval['qvalue'] <=0.05]),'\\t qval<=0.05')\n",
    "    print(len(Qval[Qval['qvalue'] <=0.01]),'\\t qval<0.01\\n')\n",
    "#   print('Original linear reg. tests counts\\n',LR1.groupby('chrom').size())\n",
    "#   print('After FDR correction test counts\\n',LR0.groupby('chrom').size())    \n",
    "    return()\n",
    "\n",
    "#fdrcorrection('Lung/SNP_Analysis')\n",
    "for T in tissu:\n",
    "    fdrcorrection(T)#+'/SNP_Analysis')\n",
    "    \n",
    "##########   Next: We run the code to prepare for heritability analysis (STR+SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "WholeBlood  variants ...\n",
    "(209131, 12)   to   (14845, 12)\n",
    "14845\n",
    "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:25: SettingWithCopyWarning: \n",
    "A value is trying to be set on a copy of a slice from a DataFrame\n",
    "\n",
    "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
    "14845  total tests...  5015  pvalues were reduced to 1\n",
    "\n",
    "630 \t qval<=0.1\n",
    "462 \t qval<=0.05\n",
    "261 \t qval<0.01\n",
    "\n",
    "Cells-Transformedfibroblasts  variants ...\n",
    "(216353, 12)   to   (15362, 12)\n",
    "15362\n",
    "15362  total tests...  4885  pvalues were reduced to 1\n",
    "\n",
    "955 \t qval<=0.1\n",
    "703 \t qval<=0.05\n",
    "395 \t qval<0.01\n",
    "\n",
    "Muscle-Skeletal  variants ...\n",
    "(215860, 12)   to   (15275, 12)\n",
    "15275\n",
    "15275  total tests...  5546  pvalues were reduced to 1\n",
    "\n",
    "304 \t qval<=0.1\n",
    "205 \t qval<=0.05\n",
    "0 \t qval<0.01\n",
    "\n",
    "Lung  variants ...\n",
    "(225784, 12)   to   (15711, 12)\n",
    "15711\n",
    "15711  total tests...  5673  pvalues were reduced to 1\n",
    "\n",
    "378 \t qval<=0.1\n",
    "264 \t qval<=0.05\n",
    "0 \t qval<0.01\n",
    "\n",
    "Adipose-Subcutaneous  variants ...\n",
    "(225770, 12)   to   (15918, 12)\n",
    "15918\n",
    "15918  total tests...  5751  pvalues were reduced to 1\n",
    "\n",
    "376 \t qval<=0.1\n",
    "287 \t qval<=0.05\n",
    "115 \t qval<0.01\n",
    "\n",
    "Artery-Tibial  variants ...\n",
    "(221263, 12)   to   (15584, 12)\n",
    "15584\n",
    "15584  total tests...  5360  pvalues were reduced to 1\n",
    "\n",
    "524 \t qval<=0.1\n",
    "383 \t qval<=0.05\n",
    "196 \t qval<0.01\n",
    "\n",
    "Esophagus-Mucosa  variants ...\n",
    "(227378, 12)   to   (16111, 12)\n",
    "16111\n",
    "16111  total tests...  5582  pvalues were reduced to 1\n",
    "\n",
    "475 \t qval<=0.1\n",
    "330 \t qval<=0.05\n",
    "128 \t qval<0.01\n",
    "\n",
    "In [ ]:\n",
    "\n",
    "​\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
