{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.sandbox.regression.predstd import wls_prediction_std\n",
    "import subprocess\n",
    "\n",
    "def round1adjust(annot, linreg):\n",
    "    #Set up empty dataframe\n",
    "    index=['0']\n",
    "    Out_Lin_reg = pd.DataFrame(index=index, columns=linreg.columns)\n",
    "    Out_Lin_reg = Out_Lin_reg.fillna(0)\n",
    "    Out_Lin_reg['PVAL']=0\n",
    "    Test=[]        #Number of STRs tested for a gene\n",
    "    Notest=[]      #Gene twith no tested genes\n",
    "    for index, gene in annot.iterrows():\n",
    "        geneid = gene['gene.id']\n",
    "        start = gene['gene.start']\n",
    "        stop = gene['gene.stop']\n",
    "#        strs = linreg.loc[linreg['str.start'].isin(range(start,stop))]\n",
    "        strs = linreg.loc[linreg['gene'].isin([geneid])]\n",
    "        N = len(strs)\n",
    "#identifying the STR test with smallest pvalue\n",
    "        if N>0 :\n",
    "            ind_low = strs.loc[strs['p.wald'].idxmin()]\n",
    "            lowpval = strs.loc[[ind_low.name]]\n",
    "#adjusting by the number of tests\n",
    "            Adj_pval = lowpval['p.wald']*N\n",
    "            lowpval['PVAL'] = Adj_pval \n",
    "#Append adjusted STR in the output for final FDR adjustment\n",
    "            Out_Lin_reg.loc[len(Out_Lin_reg)] = lowpval.values.tolist()[0]\n",
    "            Test.append(N)\n",
    "    \n",
    "        else:\n",
    "            Notest.append(gene['gene.id']) \n",
    "    Out_Lin_reg = Out_Lin_reg.drop('0')\n",
    "    print(len(Notest), ' genes were not tested for eSTRs')\n",
    "    print(len(Test), ' genes were tested for eSTRs\\n') \n",
    "    return(Out_Lin_reg, Notest, Test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 13)\n",
      "326  genes were not tested for eSTRs\n",
      "1672  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr1 \t (1672, 14)  TO...  (1673, 14)\n",
      "161  genes were not tested for eSTRs\n",
      "1069  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr2 \t (1069, 14)  TO...  (2742, 14)\n",
      "139  genes were not tested for eSTRs\n",
      "902  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr3 \t (902, 14)  TO...  (3644, 14)\n",
      "118  genes were not tested for eSTRs\n",
      "611  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr4 \t (611, 14)  TO...  (4255, 14)\n",
      "123  genes were not tested for eSTRs\n",
      "737  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr5 \t (737, 14)  TO...  (4992, 14)\n",
      "171  genes were not tested for eSTRs\n",
      "828  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr6 \t (828, 14)  TO...  (5820, 14)\n",
      "132  genes were not tested for eSTRs\n",
      "754  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr7 \t (754, 14)  TO...  (6574, 14)\n",
      "109  genes were not tested for eSTRs\n",
      "556  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr8 \t (556, 14)  TO...  (7130, 14)\n",
      "119  genes were not tested for eSTRs\n",
      "648  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr9 \t (648, 14)  TO...  (7778, 14)\n",
      "128  genes were not tested for eSTRs\n",
      "617  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr10 \t (617, 14)  TO...  (8395, 14)\n",
      "321  genes were not tested for eSTRs\n",
      "924  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr11 \t (924, 14)  TO...  (9319, 14)\n",
      "153  genes were not tested for eSTRs\n",
      "866  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr12 \t (866, 14)  TO...  (10185, 14)\n",
      "45  genes were not tested for eSTRs\n",
      "270  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr13 \t (270, 14)  TO...  (10455, 14)\n",
      "104  genes were not tested for eSTRs\n",
      "499  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr14 \t (499, 14)  TO...  (10954, 14)\n",
      "85  genes were not tested for eSTRs\n",
      "505  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr15 \t (505, 14)  TO...  (11459, 14)\n",
      "101  genes were not tested for eSTRs\n",
      "726  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr16 \t (726, 14)  TO...  (12185, 14)\n",
      "181  genes were not tested for eSTRs\n",
      "949  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr17 \t (949, 14)  TO...  (13134, 14)\n",
      "47  genes were not tested for eSTRs\n",
      "230  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr18 \t (230, 14)  TO...  (13364, 14)\n",
      "885  genes were not tested for eSTRs\n",
      "493  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr19 \t (493, 14)  TO...  (13857, 14)\n",
      "104  genes were not tested for eSTRs\n",
      "419  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr20 \t (419, 14)  TO...  (14276, 14)\n",
      "69  genes were not tested for eSTRs\n",
      "154  genes were tested for eSTRs\n",
      "\n",
      "Done with  chr21 \t (154, 14)  TO...  (14430, 14)\n",
      "781  genes were not tested for eSTRs\n",
      "0  genes were tested for eSTRs\n",
      "\n",
      "Done with  chrX \t (0, 14)  TO...  (14430, 14)\n",
      "28  genes were not tested for eSTRs\n",
      "0  genes were tested for eSTRs\n",
      "\n",
      "Done with  chrY \t (0, 14)  TO...  (14430, 14)\n",
      "End\n",
      "14429  Total number of tests\n",
      "4705  Total number of test with pval reduced to 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/lib64/python3.4/site-packages/pandas/core/generic.py:4702: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self._update_inplace(new_data)\n",
      "/usr/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2881: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "Input='/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/Cells-Transformedfibroblasts/'\n",
    "\n",
    "#open gene annotation\n",
    "#Looking at these columns gene.chr\tgene.start\tgene.stop\tgene.id\n",
    "Annot = pd.read_csv('~/projects/GTEX_eSTRs/data/Lin_Reg/Gene_Exp_Annotation.txt', sep=',')\n",
    "#open Linear regression output\n",
    "#Looking at these columns gene\tchrom\tstr.start\tp.wald\n",
    "Linreg = pd.read_csv(Input+'/Lin_Reg_Out', sep='\\t')\n",
    "#Set up the test set\n",
    "chrom = ['chr'+str(i) for i in range(1,22)]\n",
    "chrom.append('chrX') ; chrom.append('chrY')\n",
    "##chrom=['chr10']\n",
    "#Setting up\n",
    "index=['0']\n",
    "Adjusted = pd.DataFrame(index=index, columns=Linreg.columns)\n",
    "Adjusted = Adjusted.fillna(0)\n",
    "print (Adjusted.shape)\n",
    "NT = []\n",
    "#phase1 single out the STR with lowest pval and adjust by chromosome\n",
    "for ch in chrom:\n",
    "    A = Annot.loc[Annot['gene.chr'].isin([ch])]\n",
    "    LR = Linreg.loc[Linreg['chrom'].isin([ch])]\n",
    "#    print(ch,' ',len(A), len(LR))\n",
    "    Adjval, GTEST, NTest = round1adjust(A, LR)\n",
    "#    Adjval.drop('0')\n",
    "    Adjusted = pd.concat([Adjusted, Adjval])#Adjusted.append(Adjval, ignore_index=True)\n",
    "    NT =NT + NTest\n",
    "    print('Done with ', ch, '\\t', Adjval.shape,' TO... ',Adjusted.shape)\n",
    "print('End')\n",
    "Adjusted.to_csv('LinReg_adj.tsv', sep='\\t')\n",
    "\n",
    "\n",
    "Out_Lin_reg=Adjusted.drop('0')\n",
    "#SOmetime, the p-values multiplied by the number of STRs is >1 \n",
    "Out_Lin_reg['NTest']=NT\n",
    "Out_Lin_reg['PVAL'][Out_Lin_reg['PVAL']>1] = 1\n",
    "PVAL=Out_Lin_reg['PVAL']\n",
    "PVAL.to_csv('pvalues.txt', sep='\\n', index=False)\n",
    "print(len(PVAL), ' Total number of tests')\n",
    "print(len(PVAL[PVAL>=1]),' Total number of test with pval reduced to 1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14429, 6) \t (14429, 15)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/lib/python3.4/site-packages/ipykernel/__main__.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FDR correction summary: \n",
      "eSTRs counts\t Treshold\n",
      "1312 \t qval<=0.1\n",
      "987 \t qval<=0.05\n",
      "628 \t qval<0.01\n"
     ]
    }
   ],
   "source": [
    "#Now, we use the QVALUE package to adjust the pvalues and obtain the qvalues\n",
    "Tell = subprocess.call(\"/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/fdr-correct.r\")\n",
    "Qval=pd.read_csv('/home/szfeupe/projects/GTEX_eSTRs/gtex-estrs/Scripts/qvalues.txt', sep=' ')\n",
    "print (Qval.shape, '\\t', Out_Lin_reg.shape)\n",
    "#Add output from QVALUE\n",
    "Out_Lin_reg['pval_test']=list(Qval['pvalue'])\n",
    "Out_Lin_reg['qvalue']=list(Qval['qvalue'])\n",
    "Out_Lin_reg['significant']=list(Qval['significant'])\n",
    "#Organize and save\n",
    "col=['gene', 'chrom', 'str.id', 'str.start', 'NTest', 'p.wald', 'qvalue', 'significant', 'beta','beta.se', 'PVAL', 'pval_test', 'af.dummy', 'allele1.dummy', 'allele2.dummy','lambda.remel', 'n.miss']\n",
    "Out_Lin_reg=Out_Lin_reg[col]\n",
    "Out_Lin_reg.to_csv(Input+'PQValues.txt', sep='\\t', index=False)\n",
    "#Summarize it\n",
    "print(\"FDR correction summary: \\neSTRs counts\\t Treshold\")\n",
    "print(len(Qval[Qval['qvalue'] <=0.1]),'\\t qval<=0.1')\n",
    "print(len(Qval[Qval['qvalue'] <=0.05]),'\\t qval<=0.05')\n",
    "print(len(Qval[Qval['qvalue'] <=0.01]),'\\t qval<0.01')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "genes1=['ENSG00000115129.9','ENSG00000018280.12','ENSG00000164692.13','ENSG00000146648.11','ENSG00000111537.4','ENSG00000149948.9','ENSG00000007171.12','ENSG00000100985.7','ENSG00000100292.12']\n",
    "tes = Out_Lin_reg.loc[Out_Lin_reg['gene'].isin(genes1)]\n",
    "print (tes[['gene','p.wald','beta', 'qvalue']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plt.hist(PVAL, label='pvalues')\n",
    "plt.hist(Out_Lin_reg['qvalue'], label='qvalues')\n",
    "plt.show()\n",
    "plt.hist(Out_Lin_reg['qvalue'], label='qvalues')\n",
    "print(len(PVAL[PVAL>=1]))\n",
    "print(len(PVAL))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "2- FDR correction by chromosome\n",
    "\t1 we are performing a gene level correction\n",
    "for a given gene we count the number N of tested STRs\n",
    "we identify the STR test with smallest pvalue\n",
    "that value is then adjusted by the number of test -> pvalue x N\n",
    "then we use the QVALUE package to adjust the pvalues and obtain the qvalues\n",
    "We then pick a threshold eg 0.05 to identify genes with eSTRs\n",
    "Te we will perform the fine mapping that is where we actually make sure the STR tested in \n",
    "question are the real eSTRs and not the ones left off in the selection made above.\n",
    "    genes.append(gene['gene.id'])\n",
    "        strid.append(lowpval.loc[lowpval['str.id']])\n",
    "        Nbr_test.append(N)\n",
    "        pvalue.append(lowpval.loc[lowpval['p.wald']])\n",
    "        \n",
    "        \n",
    "        \n",
    "index=['0']\n",
    "Out_Lin_reg = pd.DataFrame(index=index, columns=linreg.columns)\n",
    "Out_Lin_reg = Out_Lin_reg.fillna(0)\n",
    "Out_Lin_reg['PVAL']=0\n",
    "Test=[]        #Number of STRs tested for a gene\n",
    "Notest=[]      #Gene twith no tested genes\n",
    "genes=[]; strid=[]; pvalue=[]; Nbr_test=[]\n",
    "#For each gene in the set, count the number N of tested STRs\n",
    "for index, gene in annot.iterrows():\n",
    "    geneid = gene['gene.id']\n",
    "    start = gene['gene.start']\n",
    "    stop = gene['gene.stop']\n",
    "#    strs = linreg.loc[linreg['str.start'].isin(range(start,stop))]\n",
    "    strs = linreg.loc[linreg['gene'].isin([geneid])]\n",
    "    N = len(strs)\n",
    "    #print (N,' ',gene['gene.id'], '\\t', len(list(strs['p.wald'])),'\\t', min(strs['p.wald']))\n",
    "    \n",
    "#identifying the STR test with smallest pvalue\n",
    "    if N>0 :\n",
    "        ind_low = strs.loc[strs['p.wald'].idxmin()]\n",
    "        lowpval = strs.loc[[ind_low.name]]\n",
    "        #print (strs['p.wald'], '\\n\\n', lowpval['p.wald'])\n",
    "#        print (N,' ',gene['gene.id'], '\\t', len(list(strs['p.wald'])),'\\t', min(strs['p.wald']))\n",
    "#adjusting by the number of tests\n",
    "        Adj_pval = lowpval['p.wald']*N\n",
    "        lowpval['PVAL'] = Adj_pval\n",
    "#        print (gene['gene.id'], '\\t', lowpval['gene'],'\\t',N)\n",
    "    \n",
    "#Append adjusted STR in the output for final FDR adjustment\n",
    "        Out_Lin_reg.loc[len(Out_Lin_reg)] = lowpval.values.tolist()[0]\n",
    "        Test.append(N)\n",
    "    \n",
    "    else:\n",
    "        Notest.append(gene['gene.id']) \n",
    "        #print(gene['gene.id'],'----------------------------------')\n",
    "\n",
    "Out_Lin_reg = Out_Lin_reg.drop('0')\n",
    "print(len(Notest), ' genes were not tested for eSTRs')\n",
    "print(len(Test), ' genes were tested for eSTRs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p = [1,2,3]\n",
    "q = [0,9,8]\n",
    "p+q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
