{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/python2.7\n",
    "\n",
    "import argparse\n",
    "import math\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import random\n",
    "import shutil\n",
    "import sys\n",
    "import gzip\n",
    "\n",
    "#sys.path.append(\"/san/melissa/workspace/str-qtl/lmm/\")\n",
    "#from LMMSimulationsUtils2 import *\n",
    "\n",
    "\"\"\"\n",
    "Analyze heritability of gene expression due to SNPs/STRs\n",
    "Notes:\n",
    "** The reported SE for STRs when treating the STR as fixed effect are the SE on *beta*, not on *beta^2* **\n",
    "\"\"\"\n",
    "\n",
    "GCTA_MIN_VE = 0.000001 # Lowest VE that GCTA reports when constrained\n",
    "\n",
    "def PROGRESS(msg, printit=True):\n",
    "    if printit: # false for some messages when not in debug mode\n",
    "        sys.stderr.write(\"%s\\n\"%msg.strip())\n",
    "\n",
    "def GetMAF(x):\n",
    "    \"\"\"\n",
    "    Get SNP MAF\n",
    "    \"\"\"\n",
    "    x=x.convert_objects(convert_numeric=True)\n",
    "    vals = x.values\n",
    "#    print(len(vals))\n",
    "    maf = sum(vals)*1.0/(2*len(vals))\n",
    "    return min([maf, 1-maf])\n",
    "\n",
    "def GRM(snpdata):\n",
    "    \"\"\"\n",
    "    Return n by n GRM, scaled to have mean diagonal element 1\n",
    "    \"\"\"\n",
    "#    snpdata.apply(pd.to_numeric, errors='coerce')\n",
    "    p, n = snpdata.shape\n",
    "    print(p,'  ',n)\n",
    "    K = np.zeros((n, n))\n",
    "    for i in range(p):\n",
    "#        print(snpdata.iloc[i,:])\n",
    "        gt = snpdata.iloc[i,:].apply(lambda x: float(x))\n",
    "        var = np.var(gt)\n",
    "        #print(var)\n",
    "        if var == 0: continue\n",
    "        m = np.mean(gt)\n",
    "        x = np.reshape((np.array(gt)-m).transpose(), (n, 1))\n",
    "        gt_m = 1/(var*p)*x.dot(x.transpose())\n",
    "        K = K + gt_m\n",
    "#    print(K )\n",
    "    # Make sure diagonal had mean 1 (should anyway)\n",
    "    diag_mean = np.mean(np.diagonal(K))\n",
    "#    print(np.diagonal(K))\n",
    "    K = K/diag_mean\n",
    "    return K\n",
    "\n",
    "def WriteGCTAPhenotypeFile(locus, exprfile):\n",
    "    \"\"\"\n",
    "    Write the phenotype file \n",
    "    \"\"\"\n",
    "    n=len(locus)\n",
    "    f = open(exprfile, \"w\")\n",
    "    for i in range(n):\n",
    "        f.write(' '.join([str(i),str(i),str(locus[i]),'\\n']))\n",
    "    f.close()\n",
    "\n",
    "def WriteGCTAGRM(K, grmfile, p):\n",
    "    \"\"\"\n",
    "    Calculate GRM and output file in GCTA format (.gz)\n",
    "    Need to write:\n",
    "    $grmfile.grm.gz: ind1, ind2, num nonmissing SNPs, relatedness (space-separated) (indices start at 1, rows into $grmfile.ind)\n",
    "      only includes lower triangle of GRM\n",
    "    $grmfile.ind: family ID, individual ID (space-separated)\n",
    "    \"\"\"\n",
    "    n = K.shape[0]\n",
    "    # Calculate GRM\n",
    "    f = gzip.open(\"%s.grm.gz\"%grmfile, \"wb\")\n",
    "    for i in range(n):\n",
    "        for j in range(i, n):\n",
    "            val = K[i, j]\n",
    "#it was     f.write(\" \".join(map(str, [j+1, i+1, p, val]))+\"\\n\")\n",
    "            f.write(bytes(\" \".join(map(str, [j+1, i+1, p, val]))+\"\\n\" , 'UTF-8'))\n",
    "    f.close()\n",
    "    # Write ind file\n",
    "    f = open(\"%s.grm.id\"%grmfile, \"w\")\n",
    "    for i in range(n):\n",
    "        f.write(\" \".join(map(str, [i, i]))+\"\\n\")\n",
    "    f.close()\n",
    "\n",
    "def ParseGCTAResults(gctafile, include_str):\n",
    "    \"\"\"\n",
    "    return cis_snp_h2, cis_snp_h2_se, cis_str_h2, cis_str_h2_se, logL\n",
    "    \"\"\"\n",
    "    f = open(gctafile, \"r\")\n",
    "    lines = f.readlines()\n",
    "    cis_str_h2 = None\n",
    "    cis_str_h2_se = None\n",
    "    for line in lines:\n",
    "        items = line.strip().split()\n",
    "        if len(items) < 1: continue\n",
    "        if include_str == \"NO\" or include_str == \"FE\" or include_str == \"SAMPLES\":\n",
    "            if items[0] == \"V(G)\":\n",
    "                cis_snp_h2 = items[1]\n",
    "                cis_snp_h2_se = items[2]\n",
    "        else:\n",
    "            if items[0] == \"V(G1)\":\n",
    "                cis_snp_h2 = items[1]\n",
    "                cis_snp_h2_se = items[2]\n",
    "            if items[0] == \"V(G2)\":\n",
    "                cis_str_h2 = items[1]\n",
    "                cis_str_h2_se = items[2]\n",
    "        if items[0] == \"logL\":\n",
    "            logL = items[1]\n",
    "        if include_str == \"FE\":\n",
    "            if items[0] == \"Fix_eff\":\n",
    "                ind = lines.index(line)+2\n",
    "                items = lines[ind].strip().split()\n",
    "                cis_str_h2 = float(items[0])**2\n",
    "                cis_str_h2_se = items[1]\n",
    "        if items[0] == \"Pval\":\n",
    "            Pval = items[1]\n",
    "    if include_str == \"FE\":\n",
    "        return cis_snp_h2, cis_snp_h2_se, cis_str_h2, cis_str_h2_se, logL, Pval\n",
    "    else:\n",
    "        return cis_snp_h2, cis_snp_h2_se, cis_str_h2, cis_str_h2_se, logL, 'N/A'\n",
    "\n",
    "def GetPermutedLocusSTRs(locus_str):\n",
    "    \"\"\"\n",
    "    Return locus_str dataframe with STR genotypes permuted\n",
    "    \"\"\"\n",
    "    gts = list(locus_str.iloc[:,0])\n",
    "    random.shuffle(gts)\n",
    "    locus_str_perm = pd.DataFrame({locus_str.columns[0]: gts})\n",
    "    locus_str_perm.index = locus_str.index\n",
    "    return locus_str_perm\n",
    "\n",
    "def z(vals):\n",
    "    vals = list(map(float, list(vals)))\n",
    "    m = np.mean(vals)\n",
    "    s = math.sqrt(np.var(vals))\n",
    "    return [(item-m)*1.0/s for item in vals]\n",
    "\n",
    "def ZNorm(locus_vars):\n",
    "    \"\"\"\n",
    "    Znormalize variants\n",
    "    \"\"\"\n",
    "    columns = locus_vars.columns\n",
    "    for c in columns:\n",
    "        locus_vars[c] = z(locus_vars[c])\n",
    "    return locus_vars\n",
    "\n",
    "\n",
    "def WriteGCTACovarFile(locus, strcovarfile):\n",
    "    \"\"\" Write GCTA covariable file using normalized genotype\n",
    "    \"\"\"\n",
    "    f = open(strcovarfile, \"w\")\n",
    "    n=locus.shape[0]\n",
    "    for i in range(n):\n",
    "        N_geno = list(locus.iloc[i,:].values)\n",
    "        f.write(\" \".join([str(i), str(i)]+[str(m) for m in N_geno])+\"\\n\")\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "    EXPRFILE = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/WholeBlood/Corr_Expr.csv\"\n",
    "    EXPRANNOTFILE = \"~/projects/GTEX_eSTRs/data/Lin_Reg/Gene_Exp_Annotation.txt\"\n",
    "    CHROM = \"chr22\"\n",
    "    REML_NO_CONSTRAIN=True\n",
    "    if \"chr\" not in str(CHROM): CHROM=\"chr%s\"%CHROM\n",
    "    DISTFROMGENE = 10000\n",
    "    STRGTFILE = \"/storage/szfeupe/Runs/GTEx_estr/Normalized_Genotypes/STR_Norm_Gen.chr22\"\n",
    "    SNPGTFILE = \"/storage/szfeupe/Runs/GTEx_estr/SNP_Analysis/rawnormgen1\"\n",
    "    OUTFILE = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/WholeBlood/HH/Testingcenter/h2_gtca\"\n",
    "    TMPDIR = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/WholeBlood/HH/Testingcenter/tmp\"\n",
    "    SNPMAF = 0.05\n",
    "    LMM_METHOD = \"GCTA\"\n",
    "    INCLUDE_STR = \"FE\"\n",
    "    ESTR_RESULTS_FILE = \"/storage/szfeupe/Runs/GTEx_estr/Analysis_by_Tissue/WholeBlood/PQValue.tsv\"\n",
    "    ESTR_GENES_ONLY = 0.1\n",
    "    UNLINKED_CTRL=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (3,5,6,7,8,9,10,11,12,13,14,15,17,18,20,23,24,25,26,28,29,30,31,32,33,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,57,59,61,63,65,66,69,70,71,72,74,75,76,77,78,79,80,81,82,83,85,86,88,89,92,94,95,96,97,98,99,100,102,105,106,110,112,113,116,117,118,119,124,125,126,128,129,130,131,134,136,138,141,142,143,144,145,146,147,148) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "/usr/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (6,30,47,80,144) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "# Load expression and annotation\n",
    "expr = pd.read_csv(EXPRFILE)\n",
    "\n",
    "#\n",
    "expr_annot = pd.read_csv(EXPRANNOTFILE)\n",
    "expr_annot.index = expr_annot[\"probe.id\"].values\n",
    "expr_annot = expr_annot[expr_annot[\"gene.chr\"] == CHROM]\n",
    "#\n",
    "expr_annot = expr_annot.loc[[ge for ge in expr_annot.index if ge in list(expr.columns)]]\n",
    "\n",
    "# Load SNP genotypes\n",
    "snpgt = pd.read_csv(SNPGTFILE, sep=\"\\t\", header=None)\n",
    "##########TODOremove the header=None above and the next 3 lines\n",
    "header=\"chrom     start   GTEX-QV31  GTEX-R55E       GTEX-X4XY       GTEX-X62O       GTEX-N7MS       GTEX-N7MT       GTEX-NFK9       GTEX-NL3H       GTEX-NL4W       GTEX-NPJ7 GTEX-NPJ8       GTEX-O5YT       GTEX-O5YV       GTEX-O5YW       GTEX-OHPK       GTEX-OHPM       GTEX-OHPN       GTEX-OIZF       GTEX-OIZG       GTEX-OIZH       GTEX-OIZI    GTEX-OOBK       GTEX-OXRK       GTEX-OXRL       GTEX-OXRN       GTEX-OXRO       GTEX-OXRP       GTEX-P44H       GTEX-P4PP       GTEX-P4PQ       GTEX-P4QR       GTEX-P4QS     GTEX-P4QT       GTEX-P78B       GTEX-PLZ4       GTEX-PLZ5       GTEX-PLZ6       GTEX-POMQ       GTEX-POYW       GTEX-PSDG       GTEX-PVOW       GTEX-PW2O       GTEX-PWCY    GTEX-PWN1       GTEX-PWO3       GTEX-PWOO       GTEX-Q2AG       GTEX-Q2AH       GTEX-Q2AI       GTEX-Q734       GTEX-QEG4       GTEX-QEG5       GTEX-QESD       GTEX-QLQW    GTEX-QMRM       GTEX-QV44       GTEX-QVUS       GTEX-QXCU       GTEX-R3RS       GTEX-R55C       GTEX-R55G       GTEX-REY6       GTEX-RM2N       GTEX-RN64       GTEX-RNOR     GTEX-RU1J       GTEX-RU72       GTEX-RUSQ       GTEX-RWS6       GTEX-RWSA       GTEX-S32W       GTEX-S341       GTEX-S4P3       GTEX-S4Q7       GTEX-S4UY       GTEX-S4Z8    GTEX-S7PM       GTEX-S7SE       GTEX-S95S       GTEX-SIU7       GTEX-SIU8       GTEX-SN8G       GTEX-SNOS       GTEX-SSA3       GTEX-SUCS       GTEX-T2IS       GTEX-T2YK  GTEX-T5JC       GTEX-T5JW       GTEX-T6MN       GTEX-T6MO       GTEX-T8EM       GTEX-TKQ2       GTEX-TML8       GTEX-TMMY       GTEX-TMZS       GTEX-TSE9       GTEX-U3ZG    GTEX-U3ZH       GTEX-U3ZN       GTEX-U4B1       GTEX-U8XE       GTEX-UJHI       GTEX-UJMC       GTEX-UPIC       GTEX-UPJH       GTEX-UPK5       GTEX-UTHO       GTEX-V955     GTEX-VJYA       GTEX-VUSH       GTEX-W5WG       GTEX-W5X1       GTEX-WCDI       GTEX-WFG7       GTEX-WFJO       GTEX-WFON       GTEX-WHPG       GTEX-WHSB       GTEX-WHWD     GTEX-WK11       GTEX-WL46       GTEX-WOFM       GTEX-WVLH       GTEX-WWYW       GTEX-WXYG       GTEX-WY7C       GTEX-WYJK       GTEX-WZTO       GTEX-X261       GTEX-X3Y1  GTEX-X4EO       GTEX-X4XX       GTEX-X585       GTEX-X8HC       GTEX-XBEC       GTEX-XBED       GTEX-XGQ4       GTEX-XMK1       GTEX-XOT4       GTEX-XOTO       GTEX-XPT6 GTEX-XQ3S       GTEX-XUJ4       GTEX-XUW1       GTEX-XV7Q       GTEX-XXEK       GTEX-XYKS\"     \n",
    "header = header.split()\n",
    "snpgt.columns=header\n",
    "##########END TODO\n",
    "\n",
    "# Load STR genotypes\n",
    "strgt = pd.read_csv(STRGTFILE, sep=\"\\t\")\n",
    "\n",
    "# Restrict to STR samples\n",
    "str_samples = list(set(strgt.columns[2:].values).intersection(set(snpgt.columns[2:].values)))\n",
    "expr = expr.loc[str_samples,:]\n",
    "snpgt = snpgt[[\"chrom\",\"start\"] + str_samples]\n",
    "strgt = strgt[[\"chrom\",\"start\"] + str_samples]\n",
    "\n",
    "# Load STR results\n",
    "if ESTR_RESULTS_FILE is not None:\n",
    "    estr_results = pd.read_csv(ESTR_RESULTS_FILE, sep='\\t')\n",
    "    \n",
    "# Set up output file\n",
    "outf = open(OUTFILE, \"w\")\n",
    "if INCLUDE_STR == \"NO\" or INCLUDE_STR == \"SAMPLES\":\n",
    "    outf.write(\"\\t\".join([\"chrom\",\"gene\",\"num_snps\",\"cis_snp_h2\",\"cis_snp_h2_se\",\"logL\",\"nsamp\"])+\"\\n\")\n",
    "    print(0)\n",
    "elif INCLUDE_STR == \"FE\":\n",
    "    outf.write(\"\\t\".join([\"chrom\",\"gene\",\"str_start\",\"num_snps\",\"cis_snp_h2\",\"cis_snp_h2_se\",\"cis_str_h2\", \"cis_str_h2_se\", \"logL\",\"pval\"])+\"\\n\")\n",
    "    print(1)\n",
    "\n",
    "# Restrict to eSTR genes\n",
    "if ESTR_GENES_ONLY < 1:\n",
    "    genelist = set(estr_results[estr_results[\"qvalue\"]<=ESTR_GENES_ONLY].gene)\n",
    "    genelist = [item for item in genelist if item in expr_annot.index]\n",
    "    genelist = ['ENSG00000100292.12']\n",
    "    expr_annot = expr_annot.loc[genelist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting data for ENSG00000100292.12\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Length mismatch: Expected axis has 25 elements, new values have 1 elements",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-a4d5c13ab626>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     34\u001b[0m             \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0mlocus_str\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr_samples\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 36\u001b[0;31m         \u001b[0mlocus_str\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"STR_%s\"\u001b[0m\u001b[0;34m%\u001b[0m\u001b[0mbest_str_start\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     37\u001b[0m         \u001b[0;31m###So far we only choose the best str as fixed effect.But we should consider all cis STRs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0msamples_to_keep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mstr_samples\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr_samples\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlocus_str\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"None\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib64/python3.4/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__setattr__\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m   2755\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2756\u001b[0m             \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2757\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2758\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2759\u001b[0m             \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/src/properties.pyx\u001b[0m in \u001b[0;36mpandas.lib.AxisProperty.__set__ (pandas/lib.c:46249)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/usr/lib64/python3.4/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_set_axis\u001b[0;34m(self, axis, labels)\u001b[0m\n\u001b[1;32m    446\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    447\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_set_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 448\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    449\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib64/python3.4/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mset_axis\u001b[0;34m(self, axis, new_labels)\u001b[0m\n\u001b[1;32m   2800\u001b[0m             raise ValueError('Length mismatch: Expected axis has %d elements, '\n\u001b[1;32m   2801\u001b[0m                              \u001b[0;34m'new values have %d elements'\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2802\u001b[0;31m                              (old_len, new_len))\n\u001b[0m\u001b[1;32m   2803\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2804\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_labels\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Length mismatch: Expected axis has 25 elements, new values have 1 elements"
     ]
    }
   ],
   "source": [
    "estr_results = estr_results.loc[estr_results['gene']=='ENSG00000100292.12']#.isin(genelist)]\n",
    "expr = expr['ENSG00000100292.12']#[genelist]\n",
    "\n",
    "# For each gene, pull out data and perform specified method\n",
    "for i in range(expr_annot.shape[0]):\n",
    "    gene = 'ENSG00000100292.12' #expr_annot.index.values[i]\n",
    "    ensgene = expr_annot[\"gene.id\"].values[i]\n",
    "    print(\"Getting data for %s\"%gene)\n",
    "    genedir = os.path.join(TMPDIR,\"%s/\"%gene)\n",
    "    if not os.path.exists(genedir):\n",
    "        os.mkdir(genedir)\n",
    "    start = expr_annot[\"gene.start\"].values[i]\n",
    "    end = expr_annot[\"gene.stop\"].values[i]\n",
    "# Pull out STRs\n",
    "    samples_to_keep = str_samples\n",
    "    best_str_start = None\n",
    "    if INCLUDE_STR != \"NO\":\n",
    "        try:\n",
    "            if UNLINKED_CTRL:\n",
    "                possible_starts = list(strgt[(strgt[\"start\"] >= (start-DISTFROMGENE)) & (strgt[\"start\"] <= (end+DISTFROMGENE))].start)\n",
    "                best_str_start = random.sample(possible_starts, 1)[0]\n",
    "            else:\n",
    "# make sure to match on Ensembl gene (gene is ILMN if using array)\n",
    "                best_str_start = estr_results[estr_results[\"gene\"]==ensgene].sort(\"p.wald\")[\"str.start\"].values[0]\n",
    "        except:\n",
    "            print(\"[%s]\\tERROR: couldn't find STR LMM results\"%gene)\n",
    "            continue\n",
    "        try:\n",
    "            cis_strs = strgt[(strgt[\"start\"] >= (start-DISTFROMGENE)) & (strgt[\"start\"] <= (end+DISTFROMGENE))]\n",
    "            locus_str = cis_strs[samples_to_keep].transpose()\n",
    "#            locus_str = strgt[(strgt[\"start\"] == best_str_start)].iloc[[0],:][str_samples].transpose()\n",
    "        except:\n",
    "            print(\"[%s]\\tERROR: couldn't find STR genotypes for position %s\"%(gene, best_str_start))\n",
    "            continue\n",
    "        locus_str.index = str_samples\n",
    "        locus_str.columns = [\"STR_%s\"%best_str_start]\n",
    "        ###So far we only choose the best str as fixed effect.But we should consider all cis STRs \n",
    "        samples_to_keep = [str_samples[k] for k in range(len(str_samples)) if str(locus_str.iloc[:,0].values[k]) != \"None\"]\n",
    "        locus_str = locus_str.loc[samples_to_keep,:]\n",
    "# Make sure STRs are normalized\n",
    "        try:\n",
    "            locus_str = ZNorm(locus_str)\n",
    "        except:\n",
    "            print(\"[%s]\\tERROR: couldn't Z normalize STR genotypes\"%(gene))\n",
    "            continue\n",
    "    # Pull out SNPs\n",
    "    cis_snps = snpgt[(snpgt[\"start\"] >= (start-DISTFROMGENE)) & (snpgt[\"start\"] <= (end+DISTFROMGENE))]\n",
    "    locus_snp = cis_snps[samples_to_keep].transpose()\n",
    "    locus_snp.index = samples_to_keep\n",
    "    locus_snp = locus_snp.dropna(axis=1, how='any')\n",
    "    locus_snp.columns = cis_snps[\"start\"].apply(lambda x: \"SNP_%s\"%x)\n",
    "    #locus_snp_maf = locus_snp.apply(lambda x: GetMAF(x), 0)\n",
    "    #print(locus_snp)\n",
    "#    \n",
    "#    if len(locus_snp_maf) == 0:\n",
    "#        continue    \n",
    "#    \n",
    "#    locus_snp = locus_snp.loc[:,[i for i in range(len(locus_snp_maf)) if locus_snp_maf[i]>=SNPMAF]]\n",
    "    if locus_snp.shape[1] == 0:\n",
    "        print(\"[%s]\\tERROR: no common SNPs in region\"%gene)\n",
    "        continue\n",
    "# Get expression\n",
    "    y = pd.DataFrame({\"expr\":list(expr.loc[:,gene])})\n",
    "    y.index = str_samples\n",
    "    locus_y = y.loc[samples_to_keep,[\"expr\"]]\n",
    "# Z normalize\n",
    "    locus_y = (locus_y - np.mean(locus_y))/math.sqrt(np.var(locus_y))\n",
    "# Make SNP GRM\n",
    "    locus_snp=locus_snp.apply(pd.to_numeric, errors='coerce')\n",
    "    locus_snp = locus_snp.dropna(axis=1, how='any')\n",
    "    K = GRM(locus_snp.transpose())\n",
    "    if str(np.mean(K)) == \"nan\":\n",
    "        print(\"[%s]\\tERROR: nans in GRM\"%gene)\n",
    "\n",
    "# Write GRM\n",
    "    if LMM_METHOD == \"GCTA\":\n",
    "        exprfile = os.path.join(genedir, \"expr.pheno\")\n",
    "        mgrmfile = os.path.join(genedir, \"mgrm.txt\")\n",
    "        snpgrmfile = os.path.join(genedir, \"snp.grm.txt\")\n",
    "        if REML_NO_CONSTRAIN:\n",
    "            reml_command = \"--reml-no-constrain\"\n",
    "        else: reml_command = \"--reml\"\n",
    "        gcta_cmd = \"/storage/resources/source/gcta64 %s --mgrm-gz %s --pheno %s --out %s/gcta \"%(reml_command, mgrmfile, exprfile, genedir)\n",
    "        g = open(mgrmfile, \"w\")\n",
    "        g.write(snpgrmfile+\"\\n\")\n",
    "        WriteGCTAGRM(K, snpgrmfile, p=locus_snp.shape[1])\n",
    "        if INCLUDE_STR == \"FE\": # --qcovar\n",
    "            strcovarfile = os.path.join(genedir, \"str.qcovar\")\n",
    "            WriteGCTACovarFile(locus_str, strcovarfile)\n",
    "            gcta_cmd += \" --qcovar %s --reml-est-fix\"%strcovarfile\n",
    "        if INCLUDE_STR == \"RE\":\n",
    "            K_str = GRM(locus_str.transpose())\n",
    "            strgrmfile = os.path.join(genedir, \"str.grm.txt\")\n",
    "            g.write(strgrmfile+\"\\n\")\n",
    "            WriteGCTAGRM(K_str, strgrmfile, p=locus_str.shape[1])\n",
    "        g.close()\n",
    "        locus_y[\"expr\"].fillna(\"NA\", inplace=True)\n",
    "        WriteGCTAPhenotypeFile(locus_y[\"expr\"].values, exprfile)\n",
    "        gcta_cmd += \" > /dev/null 2>&1\"\n",
    "        os.system(gcta_cmd)\n",
    "# Parse results\n",
    "        gctafile = os.path.join(genedir, \"gcta.hsq\")\n",
    "        if not os.path.exists(gctafile):\n",
    "            print(\"[%s]\\tERROR: GCTA could not analyze this gene\"%gene)\n",
    "            continue\n",
    "        cis_snp_h2, cis_snp_h2_se, cis_str_h2, cis_str_h2_se, logL, pval = ParseGCTAResults(gctafile, INCLUDE_STR)\n",
    "        # Output results\n",
    "        if INCLUDE_STR == \"NO\" or INCLUDE_STR == \"SAMPLES\":\n",
    "            word=\"\\t\".join(map(str, [CHROM, gene, locus_snp.shape[1], cis_snp_h2, cis_snp_h2_se, logL, len(samples_to_keep)]))\n",
    "            print(word)\n",
    "            #outf.write(word+\"\\n\")\n",
    "        else:\n",
    "            #outf.write(\"\\t\".join(map(str, [CHROM, gene, best_str_start, locus_snp.shape[1], cis_snp_h2, cis_snp_h2_se,\\\n",
    "            #                                   cis_str_h2, cis_str_h2_se, logL, len(samples_to_keep), pval]))+\"\\n\")  #Figure out this , len(cis_str_h2_null)       \n",
    "            print('He')\n",
    "        break\n",
    "outf.close()       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "locus_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "see=pd.read_csv('/storage/szfeupe/Runs/GTEx_estr/SNP_Analysis/SNP_Norm_Raw_Genotypes.txt', sep='\\t')\n",
    "#see['ID']='str_'+see[1].astype(str)\n",
    "see.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "estr_results[estr_results[\"gene\"]==ensgene].sort(\"p.wald\")[\"str.start\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
